{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Microbial Genomics: Lab 4\n",
    "## Topic: Bash, Genome Assembly & Comparison\n",
    "#### Tools used: Fastp, FastQC, Samtools, SPAdes, Bandage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part A: Lab Exercises\n",
    "### Exercise 1: FastQ Basics\n",
    "The FastQ format is used to contain raw DNA reads that come directly from a sequencer; these reads need to be processed in a method called assembly in order to be compiled into a usable format, such as the `fasta` files we have dealt with so far. Unlike our `fasta` files, `fastq` files have no context on where they came from or what organism they belong to- it's up to us to figure this out via assembly. Fortunately, there are many excellent tools to help us!\n",
    "\n",
    "The anatomy of a `fastq` file differs from a `fasta` file in several key ways:\n",
    "* Each DNA strand (i.e., forward and reverse) will have a separate set of reads; these are usually contained in two separate files, but can sometimes be \"interleaved\" in a single file\n",
    "* FastQ files are typically much larger than their FastA counterparts; this is because it takes many reads to create a single high-quality sequence\n",
    "* Each read in a FastQ typically contains four lines; the first begins with an @ symbol and a sequence ID, followed by the sequence itself, a line with a + symbol, and finally, a line describing the quality of the read. For example:\n",
    "\n",
    "`@SEQ_ID\n",
    "GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT\n",
    "+\n",
    "!''*((((***+))%%%++)(%%%%).1***-+*''))**55CCF>>>>>>CCCCCCC65`\n",
    "\n",
    "**Open the file ec_k12_mg1655_r1.fastq in a text editor and answer the following questions in comments below:**\n",
    "1. Based on the sequence ID, what can you tell about the sequences in this file?\n",
    "2. Pick a sequence and look at the quality line. Use [this page](https://en.wikipedia.org/wiki/FASTQ_format) to qualify the quality of the read. Is it generally high-quality?\n",
    "3. Is this file interleaved or single-stranded?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Exercise 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 2: FastQ Coverage and QC\n",
    "One of the reasons FastQ files are much larger than the genomes they represent is because high-quality reads typically have very high \"coverage\" of the genome, meaning that for each base in the full genome, a FastQ file should have upwards of 300 reads (i.e., 300x coverage); that is, for E. Coli, which has a genome size of 4.6Mb, there can be more than 1.3 _billion_ bases. \n",
    "\n",
    "Because even modestly sized genomes require FastQ files that are several GB, we'll be working with plasmid sequences, which are smaller (~2000 bp). In this exercise, we'll be performing QC (quality control) for these sequences.\n",
    "\n",
    "`fastp` is a general-purpose FastQ processing toolkit that includes QC, trimming, filtering, deduplication, and other useful functionality. For the most part, we will be concerned with qc (removing low-quality reads) and trimming (removing adapter sequences) for our pipelines. `fastqc` is another tool used to generate nice QC reports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read1 before filtering:\n",
      "total reads: 22720100\n",
      "total bases: 2272010000\n",
      "Q20 bases: 2173528270(95.6654%)\n",
      "Q30 bases: 2069237662(91.0752%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 22720100\n",
      "total bases: 2317450200\n",
      "Q20 bases: 2217453003(95.685%)\n",
      "Q30 bases: 2099886322(90.6119%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 21810827\n",
      "total bases: 2180618868\n",
      "Q20 bases: 2119614947(97.2024%)\n",
      "Q30 bases: 2024715978(92.8505%)\n",
      "\n",
      "Read2 aftering filtering:\n",
      "total reads: 21810827\n",
      "total bases: 2224204532\n",
      "Q20 bases: 2170542617(97.5874%)\n",
      "Q30 bases: 2063756635(92.7863%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 43621654\n",
      "reads failed due to low quality: 1809904\n",
      "reads failed due to too many N: 8642\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 36808\n",
      "bases trimmed due to adapters: 978926\n",
      "\n",
      "Duplication rate: 3.53444%\n",
      "\n",
      "Insert size peak (evaluated by paired-end reads): 156\n",
      "\n",
      "JSON report: fastp.json\n",
      "HTML report: lab4/fastp_report.html\n",
      "\n",
      "fastp -n 25 -h lab4/fastp_report.html -i lab4/read_1.fastq.gz -I lab4/read_2.fastq.gz -o lab4/read_1.trimmed.fastq.gz -O lab4/read_2.trimmed.fastq.gz \n",
      "fastp v0.20.1, time used: 168 seconds\n"
     ]
    }
   ],
   "source": [
    "# run FastP on our fastq files; discard reads of quality <25, trim the adapters, and generate a report\n",
    "! fastp -n 25 -h lab4/fastp_report.html -i lab4/read_1.fastq.gz -I lab4/read_2.fastq.gz -o lab4/read_1.trimmed.fastq.gz -O lab4/read_2.trimmed.fastq.gz\n",
    "\n",
    "# run fastQC on our reads and move the results to the lab4 folder\n",
    "! fastqc lab4/*.fastq.gz\n",
    "! mv *fastqc* lab4/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Using the analysis performed above, answer the following questions:**\n",
    "1. If we know that the genome that was sequenced has a length of 4.6 Mbp, and we used a sequencer with read length of 100, what is the estimated coverage for this fastq file?\n",
    "2. Open one of the fastQC HTML files. Based on this [QC tutorial](https://rtsf.natsci.msu.edu/genomics/tech-notes/fastqc-tutorial-and-faq/), are our reads high quality or not? Are there any graphs that are particularly interesting?\n",
    "3. Open up fastp_report.html. How many reads did we lose by performing QC and trimming? Does this seem like a large number?\n",
    "4. Are both reads approximately the same in terms of quality and resulting size? If not, what is different?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Exercise 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 3: Using Command Line Tools and Reference Genomes\n",
    "You may have noticed that we used a `!` above to run the `fastp` command. We're actually running commands on the command line when we do this; as we use more tools throughout this course, you'll notice that although many of them are written in Python, they exist as standalone executables that need to be called via Bash. This is because most Bioinformatics pipelines are built using pipelines of many different programs linked together, and creating such pipelines is a relatively natrual process within a shell environment.\n",
    "\n",
    "_If you haven't used Bash before, now might be a good time to review tutorials such as [this one](https://towardsdatascience.com/basics-of-bash-for-beginners-92e53a4c117a)._\n",
    "\n",
    "From Jupyter, you can run any command that you'd have access to from your Conda environment by using `!`. `samtools` is a common suite of tools to work with SAM (non-binary) and BAM (binary) files, which are raw versions of aligned reads. Below, we'll use `samtools` to start exploring our `fastq` files and calculate coverage programmatically. The steps we'll take are:\n",
    "1. Align our FastQ reads to a known reference genome\n",
    "2. Correct any mate-pair issues that happened during alignment\n",
    "3. Sort and index the resulting SAM file for fast downstream operations\n",
    "5. Check the coverage of the resulting SAM file\n",
    "\n",
    "*Note: this is a subset of a very common samtools workflow, which is covered in more depth [here](http://www.htslib.org/workflow/#fastq_to_bam)*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# run minimap to align our fastq reads to the reference genome\n",
    "! minimap2 -a -x sr lab4/mg1655_ref.fasta lab4/*_1.fastq.gz lab4/*_2.fastq.gz -o lab4/mg1655.sam\n",
    "\n",
    "# fix the mate-pairs that may have been effected by alignment\n",
    "! samtools fixmate -O sam,level=1 lab4/mg1655.sam lab4/mg1655_fixmate.sam\n",
    "\n",
    "# Sort the SAM file- this is required for coverage calculation\n",
    "! samtools sort -l 9 -o lab4/mg1655_sort.bam -T lab4/tmp/mg1655 lab4/mg1655_fixmate.sam\n",
    "\n",
    "# calculate coverage and depth of the SAM file\n",
    "! samtools coverage -o lab4/coverage.txt lab4/mg1655_sort.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Use the files generated in the above cells to answer the following questions:**\n",
    "1. Open one of the generated SAM files. How is it different from the raw fastq files?\n",
    "2. What's different between the SAM and BAM files?\n",
    "3. Take a look at the coverage.txt file. What is it telling us? \n",
    "4. The last line of the output above is the average coverage calculated by samtools. How does it compare to our estimated coverage above?\n",
    "\n",
    "*Note: you may want to delete some of the files generated above after you're done with these questions if you're running low on disk space!*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exercise 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 4: Piping & I/O\n",
    "Above, we ran several lines, one at a time, in order to produce an output. This worked, but we ended up with several (large) intermediate files. Often, we don't need to use these files, so to get around having to write them out and then read them back in, we can use bash _piping_. Essentially, this just means we take the output from one command and feed it into the next command without ever writing it into a file. This will be useful when we begin building larger pipelines below.\n",
    "\n",
    "Let's take the commands from Exercise 3 and put them into a single command. We'll use `%%bash` at the top of the cell to turn the entire cell into a bash interpreter, instead of using `!`. We'll also use the `\\` character to continue lines, for readability.\n",
    "\n",
    "_Note: One important operator we're not using explicitly here is `>`, which is used to redirect output. Using `>` at any stage in this pipeline would direct the output of the preceding command to a file, i.e., writing `cat lab4/mg1655_ref.fasta > lab4/mg1655_ref2.fasta` would take the contents of `mg1655_ref.fasta` and write them into `mg1655_ref2.fasta`. You may need to use `>` in your homework assignment, so read up on it in the bash tutorial above if you feel the need._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::0.161*1.02] collected minimizers\n",
      "[M::mm_idx_gen::0.182*1.25] sorted minimizers\n",
      "[M::main::0.182*1.25] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.182*1.25] mid_occ = 1000\n",
      "[M::mm_idx_stat] kmer size: 21; skip: 11; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.191*1.24] distinct minimizers: 757222 (99.27% are singletons); average occurrences: 1.021; average spacing: 6.004; total length: 4639675\n",
      "[M::worker_pipeline::3.637*3.17] mapped 495050 sequences\n",
      "[M::worker_pipeline::5.515*3.41] mapped 495050 sequences\n",
      "[M::worker_pipeline::7.468*3.54] mapped 495050 sequences\n",
      "[M::worker_pipeline::9.590*3.62] mapped 495050 sequences\n",
      "[M::worker_pipeline::11.621*3.67] mapped 495050 sequences\n",
      "[M::worker_pipeline::19.672*3.04] mapped 495050 sequences\n",
      "[M::worker_pipeline::20.349*3.01] mapped 495050 sequences\n",
      "[M::worker_pipeline::21.095*3.03] mapped 495050 sequences\n",
      "[M::worker_pipeline::23.222*3.13] mapped 495050 sequences\n",
      "[M::worker_pipeline::25.087*3.19] mapped 495050 sequences\n",
      "[M::worker_pipeline::32.941*2.96] mapped 495050 sequences\n",
      "[M::worker_pipeline::33.631*2.94] mapped 495050 sequences\n",
      "[M::worker_pipeline::34.362*2.96] mapped 495050 sequences\n",
      "[M::worker_pipeline::36.773*3.03] mapped 495050 sequences\n",
      "[M::worker_pipeline::38.771*3.08] mapped 495050 sequences\n",
      "[M::worker_pipeline::46.603*2.93] mapped 495050 sequences\n",
      "[M::worker_pipeline::47.303*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::48.065*2.93] mapped 495050 sequences\n",
      "[M::worker_pipeline::50.601*2.98] mapped 495050 sequences\n",
      "[M::worker_pipeline::52.490*3.02] mapped 495050 sequences\n",
      "[M::worker_pipeline::60.325*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::61.001*2.90] mapped 495050 sequences\n",
      "[M::worker_pipeline::61.737*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::63.985*2.95] mapped 495050 sequences\n",
      "[M::worker_pipeline::66.042*2.98] mapped 495050 sequences\n",
      "[M::worker_pipeline::74.200*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::74.897*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::75.669*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::78.083*2.93] mapped 495050 sequences\n",
      "[M::worker_pipeline::80.360*2.96] mapped 495050 sequences\n",
      "[M::worker_pipeline::88.434*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::89.148*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::89.915*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::92.236*2.92] mapped 495050 sequences\n",
      "[M::worker_pipeline::94.283*2.94] mapped 495050 sequences\n",
      "[M::worker_pipeline::102.244*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::102.997*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::103.726*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::105.829*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::107.835*2.93] mapped 495050 sequences\n",
      "[M::worker_pipeline::115.722*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::116.428*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::117.183*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::119.757*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::121.768*2.93] mapped 495050 sequences\n",
      "[M::worker_pipeline::129.930*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::130.657*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::131.417*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::133.889*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::136.017*2.92] mapped 495050 sequences\n",
      "[M::worker_pipeline::144.422*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::145.248*2.87] mapped 495050 sequences\n",
      "[M::worker_pipeline::146.156*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::148.824*2.90] mapped 495050 sequences\n",
      "[M::worker_pipeline::151.005*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::159.243*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::159.997*2.87] mapped 495050 sequences\n",
      "[M::worker_pipeline::160.806*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::163.117*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::165.114*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::167.096*2.92] mapped 495050 sequences\n",
      "[M::worker_pipeline::175.134*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::175.848*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::176.625*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::178.934*2.90] mapped 495050 sequences\n",
      "[M::worker_pipeline::180.937*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::188.723*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::189.417*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::190.161*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::192.371*2.90] mapped 495050 sequences\n",
      "[M::worker_pipeline::194.378*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::202.116*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::202.804*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::203.543*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::205.611*2.90] mapped 495050 sequences\n",
      "[M::worker_pipeline::207.536*2.91] mapped 495050 sequences\n",
      "[M::worker_pipeline::215.576*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::216.438*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::217.289*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::220.543*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::222.687*2.90] mapped 495050 sequences\n",
      "[M::worker_pipeline::230.764*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::231.486*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::232.241*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::234.803*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::237.358*2.90] mapped 495050 sequences\n",
      "[M::worker_pipeline::246.141*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::246.860*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::247.633*2.88] mapped 495050 sequences\n",
      "[M::worker_pipeline::250.440*2.89] mapped 495050 sequences\n",
      "[M::worker_pipeline::253.031*2.90] mapped 495050 sequences\n",
      "[M::worker_pipeline::260.603*2.82] mapped 390650 sequences\n",
      "[M::main] Version: 2.22-r1101\n",
      "[M::main] CMD: minimap2 -a -x sr lab4/mg1655_ref.fasta lab4/ERR022075_1.fastq.gz lab4/ERR022075_2.fastq.gz\n",
      "[M::main] Real time: 260.624 sec; CPU: 735.944 sec; Peak RSS: 0.735 GB\n",
      "[bam_sort_core] merging from 18 files and 1 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "minimap2 -a -x sr lab4/mg1655_ref.fasta lab4/*_1.fastq.gz lab4/*_2.fastq.gz  | \\\n",
    "samtools fixmate -u -m - - | \\\n",
    "samtools sort -u -T lab4/tmp/mg1655 - | \\\n",
    "samtools coverage -o lab4/coverage_piped.txt -"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Based on the code and results above, answer the following questions:**\n",
    "1. What does the `-` character do in the code above?\n",
    "2. What files were created in this workflow, compared to the one in Exercise 3?\n",
    "3. Are there any potential issues with this type of piped workflow?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exercise 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 5\n",
    "Now that we have some familiarity with Bash, lets assemble our FastQ files! De novo assembly is used if a reference is unknown or non-existant; it attempts to build scaffolds from the reads, and assembles these scaffolds into a full genome. When the reference genome is known, we perform a different type of assembly, known as mapping assembly or alignment. `samtools` is actually a low-level aligner, and can perform this operation, but many other more sophisticated programs exist- we'll cover those in next week's lab.\n",
    "\n",
    "For this laboratory, we'll use `spades` for de novo assembly. Others exist, but SPAdes is well-known and used often for short-read bacterial genome assembly. Note that **although we know the reference genome for our reads, we will pretend we do not for this exercise**; this is the basis of de novo assemblers. Although some include options for \"reference-guided\" assembly, they operate largely on the basis that we know very little about our input reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Command line: /Users/gregory.wood/anaconda3/bin/spades.py\t-1\t/Users/gregory.wood/Downloads/jupyter_notebooks/lab4/read_1.fastq.gz\t-2\t/Users/gregory.wood/Downloads/jupyter_notebooks/lab4/read_2.fastq.gz\t-o\t/Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 3.12.0\n",
      "  Python version: 3.8.8\n",
      "  OS: macOS-10.15.7-x86_64-i386-64bit\n",
      "\n",
      "Output dir: /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output\n",
      "Mode: read error correction and assembling\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Multi-cell mode (you should set '--sc' flag if input data was obtained with MDA (single-cell) technology or --meta flag if processing metagenomic dataset)\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/Users/gregory.wood/Downloads/jupyter_notebooks/lab4/read_1.fastq.gz']\n",
      "      right reads: ['/Users/gregory.wood/Downloads/jupyter_notebooks/lab4/read_2.fastq.gz']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Read error correction parameters:\n",
      "  Iterations: 1\n",
      "  PHRED offset will be auto-detected\n",
      "  Corrected reads will be compressed\n",
      "Assembly parameters:\n",
      "  k: automatic selection based on read length\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned OFF\n",
      "  MismatchCorrector will be SKIPPED\n",
      "  Coverage cutoff is turned OFF\n",
      "Other parameters:\n",
      "  Dir for temp files: /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/tmp\n",
      "  Threads: 16\n",
      "  Memory limit (in Gb): 250\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/spades.log\n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "== Running read error correction tool: /Users/gregory.wood/anaconda3/share/spades-3.12.0-2/bin/spades-hammer /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/corrected/configs/config.info\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  75)   Starting BayesHammer, built from N/A, git revision N/A\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  76)   Loading config from /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/corrected/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  78)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 250 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  86)   Trying to determine PHRED offset\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (main.cpp                  :  92)   Determined value is 33\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (hammer_tools.cpp          :  36)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (main.cpp                  : 113)   Size of aux. kmer data 24 bytes\n",
      "     === ITERATION 0 begins ===\n",
      "  0:00:00.001     4M / 4M    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 65535\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.94434 Gb\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 349525\n",
      "  0:00:00.017     6G / 6G    INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/read_1.fastq.gz\n",
      "  0:00:36.395     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 4620437 reads\n",
      "  0:01:08.571     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 9226662 reads\n",
      "  0:01:41.080     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 13848234 reads\n",
      "  0:02:10.504     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 18411956 reads\n",
      "  0:02:40.384     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 22720100 reads\n",
      "  0:02:40.386     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/read_2.fastq.gz\n",
      "  0:03:15.888     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 27132470 reads\n",
      "  0:03:48.514     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 31563010 reads\n",
      "  0:04:18.465     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 35807385 reads\n",
      "  0:04:53.921     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 40273813 reads\n",
      "  0:05:28.018     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 44682739 reads\n",
      "  0:05:33.846     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 45440200 reads\n",
      "  0:05:33.846     6G / 7G    INFO   K-mer Splitting          (kmer_data.cpp             : 112)   Total 45440200 reads processed\n",
      "  0:05:35.081    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:05:42.594    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 124241616 kmers in total.\n",
      "  0:05:42.594    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:05:43.398    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:05:50.875   100M / 7G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:05:51.625   100M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 57619088 bytes occupied (3.71013 bits per kmer).\n",
      "  0:05:51.625   100M / 7G    INFO   K-mer Counting           (kmer_data.cpp             : 356)   Arranging kmers in hash map order\n",
      "  0:06:00.543     1G / 7G    INFO    General                 (main.cpp                  : 148)   Clustering Hamming graph.\n",
      "  0:09:30.767     1G / 7G    INFO    General                 (main.cpp                  : 155)   Extracting clusters\n",
      "  0:10:30.455     1G / 7G    INFO    General                 (main.cpp                  : 167)   Clustering done. Total clusters: 17313299\n",
      "  0:10:30.501     1G / 7G    INFO   K-mer Counting           (kmer_data.cpp             : 376)   Collecting K-mer information, this takes a while.\n",
      "  0:10:31.792     3G / 7G    INFO   K-mer Counting           (kmer_data.cpp             : 382)   Processing /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/read_1.fastq.gz\n",
      "  0:15:16.668     3G / 7G    INFO   K-mer Counting           (kmer_data.cpp             : 382)   Processing /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/read_2.fastq.gz\n",
      "  0:21:14.709     3G / 7G    INFO   K-mer Counting           (kmer_data.cpp             : 389)   Collection done, postprocessing.\n",
      "  0:21:15.305     3G / 7G    INFO   K-mer Counting           (kmer_data.cpp             : 403)   There are 124241616 kmers in total. Among them 93407142 (75.1818%) are singletons.\n",
      "  0:21:15.305     3G / 7G    INFO    General                 (main.cpp                  : 173)   Subclustering Hamming graph\n",
      "  0:21:51.529     3G / 7G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 649)   Subclustering done. Total 272 non-read kmers were generated.\n",
      "  0:21:51.529     3G / 7G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 650)   Subclustering statistics:\n",
      "  0:21:51.529     3G / 7G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 651)     Total singleton hamming clusters: 8507754. Among them 773086 (9.08684%) are good\n",
      "  0:21:51.529     3G / 7G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 652)     Total singleton subclusters: 32977. Among them 32885 (99.721%) are good\n",
      "  0:21:51.529     3G / 7G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 653)     Total non-singleton subcluster centers: 9252084. Among them 9152162 (98.92%) are good\n",
      "  0:21:51.529     3G / 7G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 654)     Average size of non-trivial subcluster: 12.509 kmers\n",
      "  0:21:51.529     3G / 7G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 655)     Average number of sub-clusters per non-singleton cluster: 1.05446\n",
      "  0:21:51.529     3G / 7G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 656)     Total solid k-mers: 9958133\n",
      "  0:21:51.529     3G / 7G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 657)     Substitution probabilities: [4,4]((0.945118,0.0246571,0.0219269,0.00829841),(0.0209767,0.954403,0.0112657,0.0133545),(0.0133393,0.0112466,0.954423,0.0209913),(0.00828356,0.0219149,0.0246414,0.94516))\n",
      "  0:21:51.554     3G / 7G    INFO    General                 (main.cpp                  : 178)   Finished clustering.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:21:51.554     3G / 7G    INFO    General                 (main.cpp                  : 197)   Starting solid k-mers expansion in 12 threads.\n",
      "  0:27:40.069     3G / 7G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 0 produced 687948 new k-mers.\n",
      "  0:32:57.375     3G / 7G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 1 produced 25417 new k-mers.\n",
      "  0:38:03.865     3G / 7G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 2 produced 361 new k-mers.\n",
      "  0:43:15.719     3G / 7G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 3 produced 0 new k-mers.\n",
      "  0:43:15.719     3G / 7G    INFO    General                 (main.cpp                  : 222)   Solid k-mers finalized\n",
      "  0:43:15.719     3G / 7G    INFO    General                 (hammer_tools.cpp          : 220)   Starting read correction in 12 threads.\n",
      "  0:43:15.719     3G / 7G    INFO    General                 (hammer_tools.cpp          : 233)   Correcting pair of reads: /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/read_1.fastq.gz and /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/read_2.fastq.gz\n",
      "  0:43:22.500     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 0 of 1200000 reads.\n",
      "  0:43:35.781     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 0\n",
      "  0:43:40.721     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 0\n",
      "  0:43:46.889     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 1 of 1200000 reads.\n",
      "  0:44:07.121     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 1\n",
      "  0:44:13.227     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 1\n",
      "  0:44:19.494     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 2 of 1200000 reads.\n",
      "  0:44:32.334     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 2\n",
      "  0:44:37.546     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 2\n",
      "  0:44:43.732     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 3 of 1200000 reads.\n",
      "  0:44:56.735     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 3\n",
      "  0:45:01.823     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 3\n",
      "  0:45:07.719     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 4 of 1200000 reads.\n",
      "  0:45:20.751     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 4\n",
      "  0:45:25.924     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 4\n",
      "  0:45:31.927     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 5 of 1200000 reads.\n",
      "  0:45:44.957     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 5\n",
      "  0:45:50.070     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 5\n",
      "  0:45:56.031     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 6 of 1200000 reads.\n",
      "  0:46:08.746     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 6\n",
      "  0:46:13.989     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 6\n",
      "  0:46:19.928     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 7 of 1200000 reads.\n",
      "  0:46:32.672     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 7\n",
      "  0:46:37.962     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 7\n",
      "  0:46:44.079     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 8 of 1200000 reads.\n",
      "  0:46:57.079     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 8\n",
      "  0:47:02.499     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 8\n",
      "  0:47:08.401     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 9 of 1200000 reads.\n",
      "  0:47:21.136     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 9\n",
      "  0:47:26.254     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 9\n",
      "  0:47:32.074     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 10 of 1200000 reads.\n",
      "  0:47:45.342     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 10\n",
      "  0:47:51.188     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 10\n",
      "  0:47:57.070     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 11 of 1200000 reads.\n",
      "  0:48:09.952     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 11\n",
      "  0:48:15.274     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 11\n",
      "  0:48:21.083     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 12 of 1200000 reads.\n",
      "  0:48:34.311     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 12\n",
      "  0:48:39.566     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 12\n",
      "  0:48:45.613     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 13 of 1200000 reads.\n",
      "  0:48:58.561     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 13\n",
      "  0:49:03.961     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 13\n",
      "  0:49:09.753     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 14 of 1200000 reads.\n",
      "  0:49:22.686     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 14\n",
      "  0:49:27.630     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 14\n",
      "  0:49:33.430     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 15 of 1200000 reads.\n",
      "  0:49:46.786     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 15\n",
      "  0:49:53.498     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 15\n",
      "  0:50:00.001     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 16 of 1200000 reads.\n",
      "  0:50:13.227     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 16\n",
      "  0:50:18.371     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 16\n",
      "  0:50:24.302     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 17 of 1200000 reads.\n",
      "  0:50:37.275     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 17\n",
      "  0:50:42.645     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 17\n",
      "  0:50:48.153     4G / 7G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 18 of 1120100 reads.\n",
      "  0:51:00.417     4G / 7G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 18\n",
      "  0:51:05.166     4G / 7G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 18\n",
      "  0:51:06.158     3G / 7G    INFO    General                 (hammer_tools.cpp          : 274)   Correction done. Changed 4828400 bases in 3978928 reads.\n",
      "  0:51:06.158     3G / 7G    INFO    General                 (hammer_tools.cpp          : 275)   Failed to correct 429971 bases out of 4445357819.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:51:06.469    48M / 7G    INFO    General                 (main.cpp                  : 255)   Saving corrected dataset description to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/corrected/corrected.yaml\n",
      "  0:51:06.475    48M / 7G    INFO    General                 (main.cpp                  : 262)   All done. Exiting.\n",
      "\n",
      "== Compressing corrected reads (with gzip)\n",
      "\n",
      "== Dataset description file was created: /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/corrected/corrected.yaml\n",
      "\n",
      "\n",
      "===== Read error correction finished. \n",
      "\n",
      "\n",
      "===== Assembling started.\n",
      "\n",
      "\n",
      "== Running assembler: K21\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/K21/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 250 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/dataset.info) with K=21\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  77)   Converting reads to binary format for library #0 (takes a while)\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  78)   Converting paired reads\n",
      "  0:00:00.232   108M / 108M  INFO    General                 (binary_converter.hpp      :  93)   16384 reads processed\n",
      "  0:00:00.405   116M / 116M  INFO    General                 (binary_converter.hpp      :  93)   32768 reads processed\n",
      "  0:00:00.744   136M / 136M  INFO    General                 (binary_converter.hpp      :  93)   65536 reads processed\n",
      "  0:00:01.420   172M / 172M  INFO    General                 (binary_converter.hpp      :  93)   131072 reads processed\n",
      "  0:00:02.837   244M / 244M  INFO    General                 (binary_converter.hpp      :  93)   262144 reads processed\n",
      "  0:00:06.367   312M / 312M  INFO    General                 (binary_converter.hpp      :  93)   524288 reads processed\n",
      "  0:00:13.511   348M / 348M  INFO    General                 (binary_converter.hpp      :  93)   1048576 reads processed\n",
      "  0:00:26.596   360M / 360M  INFO    General                 (binary_converter.hpp      :  93)   2097152 reads processed\n",
      "  0:00:53.587   348M / 360M  INFO    General                 (binary_converter.hpp      :  93)   4194304 reads processed\n",
      "  0:01:47.297   336M / 360M  INFO    General                 (binary_converter.hpp      :  93)   8388608 reads processed\n",
      "  0:03:34.781   336M / 360M  INFO    General                 (binary_converter.hpp      :  93)   16777216 reads processed\n",
      "  0:04:47.982   156M / 360M  INFO    General                 (binary_converter.hpp      : 117)   22443574 reads written\n",
      "  0:04:48.052     4M / 360M  INFO    General                 (read_converter.hpp        :  87)   Converting single reads\n",
      "  0:04:48.216   152M / 360M  INFO    General                 (binary_converter.hpp      :  93)   16384 reads processed\n",
      "  0:04:48.289   152M / 360M  INFO    General                 (binary_converter.hpp      :  93)   32768 reads processed\n",
      "  0:04:48.435   160M / 360M  INFO    General                 (binary_converter.hpp      :  93)   65536 reads processed\n",
      "  0:04:48.725   176M / 360M  INFO    General                 (binary_converter.hpp      :  93)   131072 reads processed\n",
      "  0:04:49.296   204M / 360M  INFO    General                 (binary_converter.hpp      :  93)   262144 reads processed\n",
      "  0:04:49.635   204M / 360M  INFO    General                 (binary_converter.hpp      : 117)   268714 reads written\n",
      "  0:04:49.729     4M / 360M  INFO    General                 (read_converter.hpp        :  95)   Converting merged reads\n",
      "  0:04:49.818   148M / 360M  INFO    General                 (binary_converter.hpp      : 117)   0 reads written\n",
      "  0:04:49.851     4M / 360M  INFO    General                 (construction.cpp          : 111)   Max read length 102\n",
      "  0:04:49.851     4M / 360M  INFO    General                 (construction.cpp          : 117)   Average read length 98.3935\n",
      "  0:04:49.851     4M / 360M  INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:04:49.851     4M / 360M  INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 144 files using 12 threads. This might take a while.\n",
      "  0:04:49.851     4M / 360M  INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 65535\n",
      "  0:04:49.851     4M / 360M  INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.94434 Gb\n",
      "  0:04:49.851     4M / 360M  INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 466033\n",
      "  0:05:11.494     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 20670381 reads\n",
      "  0:05:27.170     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 41175675 reads\n",
      "  0:05:43.003     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 61615934 reads\n",
      "  0:05:59.623     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 81976319 reads\n",
      "  0:06:06.434     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 90311724 reads\n",
      "  0:06:06.434     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:06:07.621    48M / 7G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 90311724 reads\n",
      "  0:06:07.621    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:06:08.553    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 17057260 kmers in total.\n",
      "  0:06:08.553    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:06:08.717    48M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:06:08.718    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:06:08.718    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:06:08.718    48M / 7G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 65535\n",
      "  0:06:08.718    48M / 7G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.94314 Gb\n",
      "  0:06:08.718    48M / 7G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 349525\n",
      "  0:06:09.886     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 17057260 kmers\n",
      "  0:06:09.886     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 17057260 kmers.\n",
      "  0:06:10.049    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:06:10.547    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 16977717 kmers in total.\n",
      "  0:06:10.547    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:06:10.683    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:06:11.129    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:06:11.227    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 7880952 bytes occupied (3.71355 bits per kmer).\n",
      "  0:06:11.251    68M / 7G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:06:12.286    68M / 7G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:06:12.286    68M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
      "  0:06:12.286    68M / 7G    INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
      "  0:06:12.286    68M / 7G    INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
      "  0:06:15.422    68M / 7G    INFO   Early tip clipping       (early_simplification.hpp  : 184)   11121750 22-mers were removed by early tip clipper\n",
      "  0:06:15.422    68M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:06:15.496    68M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:06:16.314    76M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 225923 sequences extracted\n",
      "  0:06:16.543    76M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:06:16.913    76M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:06:17.004   120M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:06:17.004   120M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:06:17.004   120M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:06:17.392   128M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 7915600 bytes occupied (3.71248 bits per kmer).\n",
      "  0:06:17.423   196M / 7G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:07:51.080   196M / 7G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:07:51.413   196M / 7G    INFO    General                 (construction.cpp          : 464)   Processed 451767 edges\n",
      "  0:07:51.454   100M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:07:51.459   104M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 11\n",
      "  0:07:51.459   100M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 783\n",
      "  0:07:51.459   100M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 783. Coverage mad: 185.325\n",
      "  0:07:51.459   100M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:07:51.980   104M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:07:53.147   104M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:07:57.214   104M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:08:06.311   104M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
      "  0:08:06.312   104M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 743.382. Fitted coverage std. dev: 208.615\n",
      "  0:08:06.329   104M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.986278\n",
      "  0:08:06.329   104M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 72\n",
      "  0:08:06.329   104M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 72\n",
      "  0:08:06.329   104M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 4537027\n",
      "  0:08:06.329   100M / 7G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 743.382\n",
      "  0:08:06.329   100M / 7G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 72\n",
      "  0:08:06.329   100M / 7G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
      "  0:08:06.329   100M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:08:06.329   100M / 7G    INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:08:06.329   100M / 7G    INFO    General                 (graph_simplification.hpp  : 662)   Flanking coverage based disconnection disabled\n",
      "  0:08:06.329   100M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:08:06.341   100M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:08:06.341   100M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial tip clipper\n",
      "  0:08:06.391    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial tip clipper triggered 2510 times\n",
      "  0:08:06.391    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial ec remover\n",
      "  0:08:08.034   108M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial ec remover triggered 62777 times\n",
      "  0:08:08.034   108M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial isolated edge remover\n",
      "  0:08:08.040   108M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial isolated edge remover triggered 301 times\n",
      "  0:08:08.049   104M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:08:08.049   104M / 7G    INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:08:08.049   104M / 7G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:08:08.049   104M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:08:08.049   104M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.075   100M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1983 times\n",
      "  0:08:08.075   100M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.601   104M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 9092 times\n",
      "  0:08:08.601   104M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.761    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 4158 times\n",
      "  0:08:08.761    92M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:08:08.761    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.766    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 92 times\n",
      "  0:08:08.766    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:08:08.893    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1045 times\n",
      "  0:08:08.893    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.920    72M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 815 times\n",
      "  0:08:08.920    72M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:08:08.920    72M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.921    72M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 8 times\n",
      "  0:08:08.921    72M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.948    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 344 times\n",
      "  0:08:08.948    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.953    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 197 times\n",
      "  0:08:08.953    60M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:08:08.953    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.953    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 2 times\n",
      "  0:08:08.953    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.962    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 85 times\n",
      "  0:08:08.962    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.965    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 73 times\n",
      "  0:08:08.965    60M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:08:08.965    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.965    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:08:08.965    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.967    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 23 times\n",
      "  0:08:08.967    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.969    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 32 times\n",
      "  0:08:08.969    60M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:08:08.969    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.969    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:08:08.969    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.970    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 5 times\n",
      "  0:08:08.970    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.970    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 24 times\n",
      "  0:08:08.970    60M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:08:08.970    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.970    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:08:08.970    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.971    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 2 times\n",
      "  0:08:08.971    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.971    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 23 times\n",
      "  0:08:08.971    60M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:08:08.971    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.971    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:08:08.971    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.971    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:08:08.971    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 14 times\n",
      "  0:08:08.972    60M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 6 times\n",
      "  0:08:08.972    60M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
      "  0:08:08.972    60M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:08:08.972    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.975    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 83 times\n",
      "  0:08:08.975    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.984    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 6 times\n",
      "  0:08:08.984    60M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
      "  0:08:08.985    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:08:08.985    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 13\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:08:08.985    56M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:08:08.985    56M / 7G    INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:08:08.985    56M / 7G    INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\n",
      "  0:08:08.985    56M / 7G    INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\n",
      "  0:08:08.985    56M / 7G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:08:08.985    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:08.994    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:08:08.994    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:08:08.994    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:08:08.994    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:08:09.002    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:08:09.002    56M / 7G    INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:08:09.003    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:08:09.003    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 9 times\n",
      "  0:08:09.003    56M / 7G    INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:08:09.004    56M / 7G    INFO    General                 (simplification.cpp        : 476)   Average coverage = 767.318\n",
      "  0:08:09.004    56M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:08:09.004    56M / 7G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K21/assembly_graph_with_scaffolds.gfa\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:08:09.047    56M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K21/before_rr.fasta\n",
      "  0:08:09.127    56M / 7G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K21/assembly_graph.fastg\n",
      "  0:08:09.369    56M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K21/simplified_contigs.fasta\n",
      "  0:08:09.452    56M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K21/final_contigs.fasta\n",
      "  0:08:09.536    56M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:08:09.536    56M / 7G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K21/assembly_graph_with_scaffolds.gfa\n",
      "  0:08:09.592    56M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K21/before_rr.fasta\n",
      "  0:08:09.675    56M / 7G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K21/assembly_graph.fastg\n",
      "  0:08:09.900    56M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K21/simplified_contigs.fasta\n",
      "  0:08:09.981    56M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K21/final_contigs.fasta\n",
      "  0:08:10.059    56M / 7G    INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:08:10.080    52M / 7G    INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 8 minutes 10 seconds\n",
      "Max read length detected as 102\n",
      "\n",
      "== Running assembler: K33\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/K33/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 250 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/dataset.info) with K=33\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 102\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 98.3935\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 144 files using 12 threads. This might take a while.\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 65535\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.94434 Gb\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 233016\n",
      "  0:00:11.368     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 12347336 reads\n",
      "  0:00:21.704     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 24469971 reads\n",
      "  0:00:32.940     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 36618820 reads\n",
      "  0:00:45.041     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 48767037 reads\n",
      "  0:00:58.767     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 60830186 reads\n",
      "  0:01:08.719     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 72885059 reads\n",
      "  0:01:18.438     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 84909720 reads\n",
      "  0:01:22.681     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 90311724 reads\n",
      "  0:01:22.681     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:01:24.369    48M / 7G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 90311724 reads\n",
      "  0:01:24.369    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:01:26.004    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 17707637 kmers in total.\n",
      "  0:01:26.004    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:01:26.418    48M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:01:26.418    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:01:26.418    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:01:26.419    48M / 7G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 65535\n",
      "  0:01:26.419    48M / 7G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.94314 Gb\n",
      "  0:01:26.419    48M / 7G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 174762\n",
      "  0:01:28.163     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 17707637 kmers\n",
      "  0:01:28.163     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 17707637 kmers.\n",
      "  0:01:28.381    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:01:29.147    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 17668037 kmers in total.\n",
      "  0:01:29.147    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:01:29.454    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:01:29.921    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:01:30.171    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 8200928 bytes occupied (3.71334 bits per kmer).\n",
      "  0:01:30.196    68M / 7G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:01:31.377    68M / 7G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:01:31.377    68M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
      "  0:01:31.377    68M / 7G    INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
      "  0:01:31.378    68M / 7G    INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
      "  0:01:34.881    68M / 7G    INFO   Early tip clipping       (early_simplification.hpp  : 184)   11684834 34-mers were removed by early tip clipper\n",
      "  0:01:34.881    68M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:01:35.135    68M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:01:36.101    76M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 140207 sequences extracted\n",
      "  0:01:36.332    76M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:01:36.925    76M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:01:36.989   112M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:01:36.989   112M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:01:36.989   112M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:01:37.390   120M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 8217240 bytes occupied (3.7124 bits per kmer).\n",
      "  0:01:37.423   188M / 7G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:03:05.566   188M / 7G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:03:05.928   188M / 7G    INFO    General                 (construction.cpp          : 464)   Processed 280375 edges\n",
      "  0:03:05.973    92M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:03:05.976    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 41\n",
      "  0:03:05.976    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 653\n",
      "  0:03:05.977    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 653. Coverage mad: 157.156\n",
      "  0:03:05.977    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:03:06.303    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:03:07.051    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:03:10.109    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:03:17.691    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
      "  0:03:20.550    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 628.363. Fitted coverage std. dev: 178.474\n",
      "  0:03:20.565    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.454442\n",
      "  0:03:20.565    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 83\n",
      "  0:03:20.565    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 83\n",
      "  0:03:20.565    92M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 4540177\n",
      "  0:03:20.565    92M / 7G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 628.363\n",
      "  0:03:20.565    92M / 7G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 83\n",
      "  0:03:20.565    92M / 7G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
      "  0:03:20.565    92M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:03:20.565    92M / 7G    INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:03:20.565    92M / 7G    INFO    General                 (graph_simplification.hpp  : 662)   Flanking coverage based disconnection disabled\n",
      "  0:03:20.565    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:03:20.572    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:03:20.572    92M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial tip clipper\n",
      "  0:03:20.629    52M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial tip clipper triggered 3954 times\n",
      "  0:03:20.629    52M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial ec remover\n",
      "  0:03:21.497    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial ec remover triggered 33634 times\n",
      "  0:03:21.497    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial isolated edge remover\n",
      "  0:03:21.520    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Initial isolated edge remover triggered 5189 times\n",
      "  0:03:21.524    56M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:03:21.524    56M / 7G    INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:03:21.524    56M / 7G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:03:21.524    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:03:21.524    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:21.539    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1225 times\n",
      "  0:03:21.539    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:21.909    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 6786 times\n",
      "  0:03:21.909    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:21.987    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2297 times\n",
      "  0:03:21.987    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:03:21.987    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:21.990    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 60 times\n",
      "  0:03:21.990    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:03:22.089    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1135 times\n",
      "  0:03:22.089    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:03:22.097    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 210 times\n",
      "  0:03:22.097    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:03:22.097    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:03:22.098    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 12 times\n",
      "  0:03:22.098    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:03:22.116    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 153 times\r\n",
      "  0:03:22.116    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\r\n",
      "  0:03:22.120    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 56 times\r\n",
      "  0:03:22.120    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\r\n",
      "  0:03:22.120    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.121    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\r\n",
      "  0:03:22.121    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.124    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 31 times\r\n",
      "  0:03:22.124    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\r\n",
      "  0:03:22.125    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 11 times\r\n",
      "  0:03:22.125    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\r\n",
      "  0:03:22.125    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.125    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\r\n",
      "  0:03:22.125    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 5 times\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 4 times\r\n",
      "  0:03:22.126    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 3 times\r\n",
      "  0:03:22.126    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\r\n",
      "  0:03:22.126    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 3 times\r\n",
      "  0:03:22.127    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\r\n",
      "  0:03:22.127    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 4 times\r\n",
      "  0:03:22.127    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\r\n",
      "  0:03:22.127    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\r\n",
      "  0:03:22.127    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.128    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 31 times\r\n",
      "  0:03:22.128    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.132    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 3 times\r\n",
      "  0:03:22.132    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\r\n",
      "  0:03:22.132    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\r\n",
      "  0:03:22.132    56M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 12\r\n",
      "  0:03:22.132    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.132    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\r\n",
      "  0:03:22.132    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.132    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\r\n",
      "  0:03:22.132    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\r\n",
      "  0:03:22.132    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\r\n",
      "  0:03:22.132    56M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\r\n",
      "  0:03:22.132    56M / 7G    INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\r\n",
      "  0:03:22.132    56M / 7G    INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\r\n",
      "  0:03:22.132    56M / 7G    INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\r\n",
      "  0:03:22.132    56M / 7G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\r\n",
      "  0:03:22.132    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.133    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\r\n",
      "  0:03:22.133    56M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.135    52M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\r\n",
      "  0:03:22.135    52M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\r\n",
      "  0:03:22.135    52M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\r\n",
      "  0:03:22.135    52M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\r\n",
      "  0:03:22.138    52M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\r\n",
      "  0:03:22.138    52M / 7G    INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\r\n",
      "  0:03:22.139    52M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\r\n",
      "  0:03:22.139    52M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 16 times\r\n",
      "  0:03:22.139    52M / 7G    INFO    General                 (simplification.cpp        : 470)   Counting average coverage\r\n",
      "  0:03:22.139    52M / 7G    INFO    General                 (simplification.cpp        : 476)   Average coverage = 646.024\r\n",
      "  0:03:22.139    52M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\r\n",
      "  0:03:22.139    52M / 7G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K33/assembly_graph_with_scaffolds.gfa\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:03:22.180    52M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K33/before_rr.fasta\n",
      "  0:03:22.252    52M / 7G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K33/assembly_graph.fastg\n",
      "  0:03:22.447    56M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K33/simplified_contigs.fasta\n",
      "  0:03:22.516    52M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K33/final_contigs.fasta\n",
      "  0:03:22.588    52M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:03:22.588    52M / 7G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K33/assembly_graph_with_scaffolds.gfa\n",
      "  0:03:22.633    52M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K33/before_rr.fasta\n",
      "  0:03:22.706    52M / 7G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K33/assembly_graph.fastg\n",
      "  0:03:22.904    56M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K33/simplified_contigs.fasta\n",
      "  0:03:22.976    52M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K33/final_contigs.fasta\n",
      "  0:03:23.052    52M / 7G    INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:03:23.063    52M / 7G    INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 3 minutes 23 seconds\n",
      "\n",
      "== Running assembler: K55\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/K55/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 250 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from N/A, git revision N/A\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/dataset.info) with K=55\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  76)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.002     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 102\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 98.3935\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:00.003     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 144 files using 12 threads. This might take a while.\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 65535\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.94434 Gb\n",
      "  0:00:00.004     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 233016\n",
      "  0:00:11.415     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 18500261 reads\n",
      "  0:00:27.215     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 36744156 reads\n",
      "  0:00:38.988     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 54987973 reads\n",
      "  0:00:49.944     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 73053785 reads\n",
      "  0:01:00.396     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 90311724 reads\n",
      "  0:01:00.396     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:01:02.313    48M / 7G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 90311724 reads\n",
      "  0:01:02.313    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:01:03.662    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 17922634 kmers in total.\n",
      "  0:01:03.662    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:01:03.959    48M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:01:03.960    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:01:03.960    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:01:03.960    48M / 7G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 65535\n",
      "  0:01:03.960    48M / 7G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.94314 Gb\n",
      "  0:01:03.960    48M / 7G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 174762\n",
      "  0:01:05.793     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 17922634 kmers\n",
      "  0:01:05.793     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 17922634 kmers.\n",
      "  0:01:06.012    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:01:06.853    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 17936496 kmers in total.\n",
      "  0:01:06.853    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:01:07.154    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:01:07.696    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:01:07.893    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 8325560 bytes occupied (3.71335 bits per kmer).\n",
      "  0:01:07.918    68M / 7G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:01:09.082    68M / 7G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:01:09.082    68M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:01:09.329    68M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:01:13.148   908M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 13546988 sequences extracted\n",
      "  0:01:15.345   908M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:01:15.936   908M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:01:22.696     4G / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:01:22.696     4G / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:01:22.696     4G / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:01:23.100     4G / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 8316888 bytes occupied (3.71235 bits per kmer).\n",
      "  0:01:23.133     4G / 7G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:02:25.247     4G / 7G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:02:27.112     4G / 7G    INFO    General                 (construction.cpp          : 464)   Processed 12000000 edges\n",
      "  0:02:28.395     4G / 7G    INFO    General                 (construction.cpp          : 464)   Processed 24000000 edges\n",
      "  0:02:28.775     4G / 7G    INFO    General                 (construction.cpp          : 464)   Processed 27093965 edges\n",
      "  0:02:28.820     4G / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:02:28.822     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 38\n",
      "  0:02:28.822     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 438\n",
      "  0:02:28.822     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 438. Coverage mad: 112.678\n",
      "  0:02:28.822     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:02:29.093     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:02:29.786     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:02:32.037     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:02:37.068     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
      "  0:02:38.974     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 417.365. Fitted coverage std. dev: 123.994\n",
      "  0:02:38.983     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.291193\n",
      "  0:02:38.983     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 63\n",
      "  0:02:38.983     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 63\n",
      "  0:02:38.983     4G / 7G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 4541625\n",
      "  0:02:38.983     4G / 7G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 417.365\n",
      "  0:02:38.983     4G / 7G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 63\n",
      "  0:02:38.983     4G / 7G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 0\n",
      "  0:02:38.983     4G / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Gap Closer\n",
      "  0:02:38.983     4G / 7G    INFO    General                 (graph_pack.hpp            : 101)   Index refill\n",
      "  0:02:38.983     4G / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:02:38.983     4G / 7G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:02:38.984     4G / 7G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 65535\n",
      "  0:02:38.984     4G / 7G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.8278 Gb\n",
      "  0:02:38.984     4G / 7G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 174762\n",
      "  0:02:42.692    10G / 10G   INFO    General                 (edge_index_builders.hpp   :  77)   Processed 27093965 edges\n",
      "  0:02:42.692    10G / 10G   INFO    General                 (edge_index_builders.hpp   :  82)   Used 27093965 sequences.\n",
      "  0:02:42.904     4G / 10G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:02:43.703     4G / 10G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 17922634 kmers in total.\n",
      "  0:02:43.703     4G / 10G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:02:44.013     4G / 10G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:02:44.471     4G / 10G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:02:44.674     4G / 10G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 8319096 bytes occupied (3.71334 bits per kmer).\n",
      "  0:02:44.872     4G / 10G   INFO    General                 (edge_index_builders.hpp   : 107)   Collecting edge information from graph, this takes a while.\n",
      "  0:02:47.837     4G / 10G   INFO    General                 (edge_index.hpp            :  92)   Index refilled\n",
      "  0:02:47.837     4G / 10G   INFO    General                 (gap_closer.cpp            : 159)   Preparing shift maps\n",
      "  0:03:10.215     6G / 10G   INFO    General                 (gap_closer.cpp            : 119)   Processing paired reads (takes a while)\n",
      "  0:06:46.112     6G / 10G   INFO    General                 (gap_closer.cpp            : 138)   Used 22443574 paired reads\n",
      "  0:06:46.112     6G / 10G   INFO    General                 (gap_closer.cpp            : 140)   Merging paired indices\n",
      "  0:06:52.767     4G / 10G   INFO   GapCloser                (gap_closer.cpp            : 346)   Closing short gaps\n",
      "  0:07:42.235     4G / 10G   INFO   GapCloser                (gap_closer.cpp            : 380)   Closing short gaps complete: filled 1 gaps after checking 321 candidates\n",
      "  0:07:42.753     4G / 10G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:07:42.787     4G / 10G   INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:07:42.787     4G / 10G   INFO    General                 (graph_simplification.hpp  : 662)   Flanking coverage based disconnection disabled\n",
      "  0:07:42.787     4G / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:07:43.819     4G / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:07:43.819     4G / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial tip clipper\n",
      "  0:09:46.215   932M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Initial tip clipper triggered 9433128 times\n",
      "  0:09:46.215   932M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial ec remover\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:09:46.380   892M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Initial ec remover triggered 1562 times\n",
      "  0:09:46.380   892M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Initial isolated edge remover\n",
      "  0:09:46.923   420M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Initial isolated edge remover triggered 21137 times\n",
      "  0:09:47.005   344M / 10G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:09:47.005   344M / 10G   INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:09:47.005   344M / 10G   INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:09:47.005   344M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:09:47.006   344M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:47.057   316M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 868 times\n",
      "  0:09:47.057   316M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.141   188M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 4220 times\n",
      "  0:09:48.141   188M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.204   172M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 648 times\n",
      "  0:09:48.204   172M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:09:48.204   172M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.222   160M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 110 times\n",
      "  0:09:48.222   160M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.442   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 454 times\n",
      "  0:09:48.442   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.454   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 117 times\n",
      "  0:09:48.454   132M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:09:48.454   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.458   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 19 times\n",
      "  0:09:48.458   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.503   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 65 times\n",
      "  0:09:48.503   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.504   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 15 times\n",
      "  0:09:48.504   132M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:09:48.504   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.504   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:09:48.504   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.511   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 11 times\n",
      "  0:09:48.511   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.511   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 3 times\n",
      "  0:09:48.511   132M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:09:48.511   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.511   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:09:48.511   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.511   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:09:48.511   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.513   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 7 times\n",
      "  0:09:48.513   132M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:09:48.513   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.513   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:09:48.513   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.513   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
      "  0:09:48.513   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
      "  0:09:48.514   132M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
      "  0:09:48.514   132M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
      "  0:09:48.514   132M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
      "  0:09:48.514   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.515   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
      "  0:09:48.515   132M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:09:48.515   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.515   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:09:48.515   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.515   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:09:48.515   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.515   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
      "  0:09:48.515   132M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:09:48.515   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.516   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 19 times\n",
      "  0:09:48.516   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.518   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 3 times\n",
      "  0:09:48.518   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.518   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:09:48.518   132M / 10G   INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:09:48.518   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:09:48.518   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:09:48.518   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:09:48.518   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:09:48.518   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:09:48.518   132M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:09:48.518   132M / 10G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Gap Closer\n",
      "  0:09:48.518   132M / 10G   INFO    General                 (graph_pack.hpp            : 101)   Index refill\n",
      "  0:09:48.519   132M / 10G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:09:48.519   132M / 10G   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:09:48.519   132M / 10G   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 65535\n",
      "  0:09:48.519   132M / 10G   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.94086 Gb\n",
      "  0:09:48.519   132M / 10G   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 174762\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:09:48.889     6G / 10G   INFO    General                 (edge_index_builders.hpp   :  82)   Used 1012 sequences.\n",
      "  0:09:49.059   132M / 10G   INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:09:49.311   128M / 10G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 4557054 kmers in total.\n",
      "  0:09:49.311   128M / 10G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:09:49.467   128M / 10G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:09:49.640   132M / 10G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:09:49.711   132M / 10G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 2121304 bytes occupied (3.72399 bits per kmer).\n",
      "  0:09:49.771   240M / 10G   INFO    General                 (edge_index_builders.hpp   : 107)   Collecting edge information from graph, this takes a while.\n",
      "  0:09:50.020   240M / 10G   INFO    General                 (edge_index.hpp            :  92)   Index refilled\n",
      "  0:09:50.021   240M / 10G   INFO    General                 (gap_closer.cpp            : 159)   Preparing shift maps\n",
      "  0:09:50.022   240M / 10G   INFO    General                 (gap_closer.cpp            : 119)   Processing paired reads (takes a while)\n",
      "  0:10:08.690   236M / 10G   INFO    General                 (gap_closer.cpp            : 138)   Used 22443574 paired reads\n",
      "  0:10:08.690   236M / 10G   INFO    General                 (gap_closer.cpp            : 140)   Merging paired indices\n",
      "  0:10:08.690   236M / 10G   INFO   GapCloser                (gap_closer.cpp            : 346)   Closing short gaps\n",
      "  0:10:08.690   236M / 10G   INFO   GapCloser                (gap_closer.cpp            : 380)   Closing short gaps complete: filled 0 gaps after checking 7 candidates\n",
      "  0:10:08.691   236M / 10G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:10:08.691   236M / 10G   INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:10:08.691   236M / 10G   INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\n",
      "  0:10:08.691   236M / 10G   INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\n",
      "  0:10:08.691   236M / 10G   INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:10:08.691   236M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:10:08.691   236M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:10:08.691   236M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:10:08.692   236M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:10:08.692   236M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:10:08.692   236M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:10:08.692   236M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:10:08.693   236M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:10:08.693   236M / 10G   INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:10:08.693   236M / 10G   INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:10:08.693   236M / 10G   INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 2 times\n",
      "  0:10:08.693   236M / 10G   INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:10:08.693   236M / 10G   INFO    General                 (simplification.cpp        : 476)   Average coverage = 427.754\n",
      "  0:10:08.693   236M / 10G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Mismatch Correction\n",
      "  0:10:08.693   236M / 10G   INFO    General                 (graph_pack.hpp            : 109)   Normalizing k-mer map. Total 535956 kmers to process\n",
      "  0:10:09.018   236M / 10G   INFO    General                 (graph_pack.hpp            : 111)   Normalizing done\n",
      "  0:11:07.049   252M / 10G   INFO    General                 (mismatch_shall_not_pass.hp: 189)   Finished collecting potential mismatches positions\n",
      "  0:11:07.133   240M / 10G   INFO    General                 (mismatch_shall_not_pass.hp: 290)   All edges processed\n",
      "  0:11:07.135   240M / 10G   INFO    General                 (mismatch_correction.cpp   :  27)   Corrected 1 nucleotides\n",
      "  0:11:07.136   240M / 10G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:11:07.136   240M / 10G   INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K55/assembly_graph_with_scaffolds.gfa\n",
      "  0:11:07.172   240M / 10G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K55/before_rr.fasta\n",
      "  0:11:07.241   236M / 10G   INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K55/assembly_graph.fastg\n",
      "  0:11:07.431   240M / 10G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K55/simplified_contigs.fasta\n",
      "  0:11:07.498   236M / 10G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K55/intermediate_contigs.fasta\n",
      "  0:11:07.566   236M / 10G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Paired Information Counting\n",
      "  0:11:07.567   236M / 10G   INFO    General                 (graph_pack.hpp            : 109)   Normalizing k-mer map. Total 535962 kmers to process\n",
      "  0:11:07.890   236M / 10G   INFO    General                 (graph_pack.hpp            : 111)   Normalizing done\n",
      "  0:11:07.890   236M / 10G   INFO    General                 (pair_info_count.cpp       : 320)   Min edge length for estimation: 60713\n",
      "  0:11:07.890   236M / 10G   INFO    General                 (pair_info_count.cpp       : 331)   Estimating insert size for library #0\n",
      "  0:11:07.890   236M / 10G   INFO    General                 (pair_info_count.cpp       : 190)   Estimating insert size (takes a while)\n",
      "  0:11:07.996   444M / 10G   INFO    General                 (pair_info_count.cpp       :  39)   Selecting usual mapper\n",
      "  0:11:10.200   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 200000 reads\n",
      "  0:11:10.247   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 400000 reads\n",
      "  0:11:10.291   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 600000 reads\n",
      "  0:11:10.330   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 800000 reads\n",
      "  0:11:10.361   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1000000 reads\n",
      "  0:11:10.391   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1200000 reads\n",
      "  0:11:10.602   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 2200000 reads\n",
      "  0:11:13.044   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 4200000 reads\n",
      "  0:11:17.884   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 8400000 reads\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:11:25.732   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 16800000 reads\n",
      "  0:11:31.701   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  98)   Total 22443574 reads processed\n",
      "  0:11:31.747   444M / 10G   INFO    General                 (pair_info_count.cpp       : 209)   Edge pairs: 67108864 (rough upper limit)\n",
      "  0:11:31.747   444M / 10G   INFO    General                 (pair_info_count.cpp       : 213)   10765739 paired reads (47.968% of all) aligned to long edges\n",
      "  0:11:31.760   236M / 10G   INFO    General                 (pair_info_count.cpp       : 354)     Insert size = 505.196, deviation = 27.4939, left quantile = 472, right quantile = 544, read length = 102\n",
      "  0:11:31.859   428M / 10G   INFO    General                 (pair_info_count.cpp       : 371)   Filtering data for library #0\n",
      "  0:11:31.861   428M / 10G   INFO    General                 (pair_info_count.cpp       :  39)   Selecting usual mapper\n",
      "  0:11:34.069   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 200000 reads\n",
      "  0:11:34.097   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 400000 reads\n",
      "  0:11:34.099   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 600000 reads\n",
      "  0:11:34.111   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 800000 reads\n",
      "  0:11:34.113   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1000000 reads\n",
      "  0:11:34.116   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1200000 reads\n",
      "  0:11:34.134   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 2200000 reads\n",
      "  0:11:36.550   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 4200000 reads\n",
      "  0:11:41.271   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 8400000 reads\n",
      "  0:11:48.838   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 16800000 reads\n",
      "  0:11:54.467   428M / 10G   INFO    General                 (sequence_mapper_notifier.h:  98)   Total 22443574 reads processed\n",
      "  0:11:54.467   428M / 10G   INFO    General                 (pair_info_count.cpp       : 383)   Mapping library #0\n",
      "  0:11:54.467   428M / 10G   INFO    General                 (pair_info_count.cpp       : 385)   Mapping paired reads (takes a while)\n",
      "  0:11:54.467   428M / 10G   INFO    General                 (pair_info_count.cpp       : 289)   Left insert size quantile 472, right insert size quantile 544, filtering threshold 2, rounding threshold 0\n",
      "  0:11:54.477   440M / 10G   INFO    General                 (pair_info_count.cpp       :  39)   Selecting usual mapper\n",
      "  0:11:57.009   440M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 200000 reads\n",
      "  0:11:57.065   440M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 400000 reads\n",
      "  0:11:57.076   440M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 600000 reads\n",
      "  0:11:57.077   440M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 800000 reads\n",
      "  0:11:57.080   440M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1000000 reads\n",
      "  0:11:57.080   440M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 1200000 reads\n",
      "  0:11:57.128   440M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 2200000 reads\n",
      "  0:11:59.925   440M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 4200000 reads\n",
      "  0:12:05.510   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 8400000 reads\n",
      "  0:12:14.541   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  80)   Processed 16800000 reads\n",
      "  0:12:21.456   444M / 10G   INFO    General                 (sequence_mapper_notifier.h:  98)   Total 22443574 reads processed\n",
      "  0:12:21.473   240M / 10G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Distance Estimation\n",
      "  0:12:21.473   240M / 10G   INFO    General                 (distance_estimation.cpp   : 173)   Processing library #0\n",
      "  0:12:21.473   240M / 10G   INFO    General                 (distance_estimation.cpp   : 149)   Weight Filter Done\n",
      "  0:12:21.474   240M / 10G   INFO   DistanceEstimator        (distance_estimation.hpp   : 116)   Using SIMPLE distance estimator\n",
      "  0:12:21.484   240M / 10G   INFO    General                 (distance_estimation.cpp   :  34)   Filtering info\n",
      "  0:12:21.484   240M / 10G   INFO    General                 (pair_info_filters.hpp     : 242)   Start filtering; index size: 8768\n",
      "  0:12:21.486   240M / 10G   INFO    General                 (pair_info_filters.hpp     : 263)   Done filtering\n",
      "  0:12:21.486   240M / 10G   INFO    General                 (distance_estimation.cpp   : 156)   Refining clustered pair information\n",
      "  0:12:21.487   240M / 10G   INFO    General                 (distance_estimation.cpp   : 158)   The refining of clustered pair information has been finished\n",
      "  0:12:21.487   240M / 10G   INFO    General                 (distance_estimation.cpp   : 160)   Improving paired information\n",
      "  0:12:21.510   240M / 10G   INFO   PairInfoImprover         (pair_info_improver.hpp    : 103)   Paired info stats: missing = 1381; contradictional = 716\n",
      "  0:12:21.525   240M / 10G   INFO   PairInfoImprover         (pair_info_improver.hpp    : 103)   Paired info stats: missing = 104; contradictional = 56\n",
      "  0:12:21.525   240M / 10G   INFO    General                 (distance_estimation.cpp   : 103)   Filling scaffolding index\n",
      "  0:12:21.525   240M / 10G   INFO   DistanceEstimator        (distance_estimation.hpp   : 116)   Using SMOOTHING distance estimator\n",
      "  0:12:21.607   240M / 10G   INFO    General                 (distance_estimation.cpp   :  34)   Filtering info\n",
      "  0:12:21.607   240M / 10G   INFO    General                 (pair_info_filters.hpp     : 242)   Start filtering; index size: 1430\n",
      "  0:12:21.607   240M / 10G   INFO    General                 (pair_info_filters.hpp     : 263)   Done filtering\n",
      "  0:12:21.607   240M / 10G   INFO    General                 (distance_estimation.cpp   : 182)   Clearing raw paired index\n",
      "  0:12:21.611   240M / 10G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Repeat Resolving\n",
      "  0:12:21.611   240M / 10G   INFO    General                 (repeat_resolving.cpp      :  69)   Using Path-Extend repeat resolving\n",
      "  0:12:21.611   240M / 10G   INFO    General                 (launcher.cpp              : 477)   ExSPAnder repeat resolving tool started\n",
      "  0:12:21.613   240M / 10G   INFO    General                 (launcher.cpp              : 392)   Creating main extenders, unique edge length = 2000\n",
      "  0:12:21.613   240M / 10G   INFO    General                 (extenders_logic.cpp       : 275)   Estimated coverage of library #0 is 427.754\n",
      "  0:12:21.613   240M / 10G   INFO    General                 (extenders_logic.cpp       : 275)   Estimated coverage of library #0 is 427.754\n",
      "  0:12:21.613   240M / 10G   INFO    General                 (extenders_logic.cpp       : 472)   Using 1 paired-end library\n",
      "  0:12:21.613   240M / 10G   INFO    General                 (extenders_logic.cpp       : 473)   Using 1 paired-end scaffolding library\n",
      "  0:12:21.613   240M / 10G   INFO    General                 (extenders_logic.cpp       : 474)   Using 0 single read libraries\n",
      "  0:12:21.613   240M / 10G   INFO    General                 (launcher.cpp              : 420)   Total number of extenders is 3\n",
      "  0:12:21.614   240M / 10G   INFO    General                 (path_extender.hpp         : 885)   Processed 0 paths from 458 (0%)\n",
      "  0:12:21.677   240M / 10G   INFO    General                 (path_extender.hpp         : 885)   Processed 46 paths from 458 (10%)\n",
      "  0:12:21.704   240M / 10G   INFO    General                 (path_extender.hpp         : 885)   Processed 92 paths from 458 (20%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0:12:21.723   240M / 10G   INFO    General                 (path_extender.hpp         : 883)   Processed 128 paths from 458 (27%)\n",
      "  0:12:21.736   244M / 10G   INFO    General                 (path_extender.hpp         : 885)   Processed 138 paths from 458 (30%)\n",
      "  0:12:21.739   244M / 10G   INFO    General                 (path_extender.hpp         : 885)   Processed 184 paths from 458 (40%)\n",
      "  0:12:21.740   244M / 10G   INFO    General                 (path_extender.hpp         : 885)   Processed 230 paths from 458 (50%)\n",
      "  0:12:21.740   244M / 10G   INFO    General                 (path_extender.hpp         : 883)   Processed 256 paths from 458 (55%)\n",
      "  0:12:21.741   244M / 10G   INFO    General                 (path_extender.hpp         : 885)   Processed 276 paths from 458 (60%)\n",
      "  0:12:21.744   244M / 10G   INFO    General                 (path_extender.hpp         : 885)   Processed 322 paths from 458 (70%)\n",
      "  0:12:21.744   244M / 10G   INFO    General                 (path_extender.hpp         : 885)   Processed 368 paths from 458 (80%)\n",
      "  0:12:21.744   244M / 10G   INFO    General                 (path_extender.hpp         : 885)   Processed 414 paths from 458 (90%)\n",
      "  0:12:21.744   244M / 10G   INFO    General                 (launcher.cpp              : 234)   Finalizing paths\n",
      "  0:12:21.744   244M / 10G   INFO    General                 (launcher.cpp              : 236)   Deduplicating paths\n",
      "  0:12:21.746   244M / 10G   INFO    General                 (launcher.cpp              : 240)   Paths deduplicated\n",
      "  0:12:21.746   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 295)   Removing overlaps\n",
      "  0:12:21.746   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 298)   Sorting paths\n",
      "  0:12:21.746   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 305)   Marking overlaps\n",
      "  0:12:21.746   244M / 10G   INFO   OverlapRemover           (pe_resolver.hpp           : 130)   Marking start/end overlaps\n",
      "  0:12:21.748   244M / 10G   INFO   OverlapRemover           (pe_resolver.hpp           : 133)   Marking remaining overlaps\n",
      "  0:12:21.751   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 308)   Splitting paths\n",
      "  0:12:21.752   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 313)   Deduplicating paths\n",
      "  0:12:21.753   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 315)   Overlaps removed\n",
      "  0:12:21.753   244M / 10G   INFO    General                 (launcher.cpp              : 257)   Paths finalized\n",
      "  0:12:21.753   244M / 10G   INFO    General                 (launcher.cpp              : 427)   Closing gaps in paths\n",
      "  0:12:21.754   244M / 10G   INFO    General                 (launcher.cpp              : 455)   Gap closing completed\n",
      "  0:12:21.754   244M / 10G   INFO    General                 (launcher.cpp              : 286)   Traversing tandem repeats\n",
      "  0:12:21.756   244M / 10G   INFO    General                 (launcher.cpp              : 296)   Traversed 4 loops\n",
      "  0:12:21.756   244M / 10G   INFO    General                 (launcher.cpp              : 234)   Finalizing paths\n",
      "  0:12:21.756   244M / 10G   INFO    General                 (launcher.cpp              : 236)   Deduplicating paths\n",
      "  0:12:21.756   244M / 10G   INFO    General                 (launcher.cpp              : 240)   Paths deduplicated\n",
      "  0:12:21.756   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 295)   Removing overlaps\n",
      "  0:12:21.756   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 298)   Sorting paths\n",
      "  0:12:21.756   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 305)   Marking overlaps\n",
      "  0:12:21.756   244M / 10G   INFO   OverlapRemover           (pe_resolver.hpp           : 130)   Marking start/end overlaps\n",
      "  0:12:21.756   244M / 10G   INFO   OverlapRemover           (pe_resolver.hpp           : 133)   Marking remaining overlaps\n",
      "  0:12:21.757   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 308)   Splitting paths\n",
      "  0:12:21.757   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 313)   Deduplicating paths\n",
      "  0:12:21.757   244M / 10G   INFO   PEResolver               (pe_resolver.hpp           : 315)   Overlaps removed\n",
      "  0:12:21.757   244M / 10G   INFO    General                 (launcher.cpp              : 257)   Paths finalized\n",
      "  0:12:21.757   244M / 10G   INFO    General                 (launcher.cpp              : 529)   ExSPAnder repeat resolving tool finished\n",
      "  0:12:21.763   240M / 10G   INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:12:21.763   240M / 10G   INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K55/assembly_graph_with_scaffolds.gfa\n",
      "  0:12:21.803   240M / 10G   INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K55/before_rr.fasta\n",
      "  0:12:21.870   240M / 10G   INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K55/assembly_graph.fastg\n",
      "  0:12:22.046   240M / 10G   INFO    General                 (contig_output_stage.cpp   :  20)   Outputting FastG paths to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K55/final_contigs.paths\n",
      "  0:12:22.140   240M / 10G   INFO    General                 (contig_output_stage.cpp   :  20)   Outputting FastG paths to /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output//K55/scaffolds.paths\n",
      "  0:12:22.233   240M / 10G   INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:12:22.633    52M / 10G   INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 12 minutes 22 seconds\n",
      "\n",
      "===== Assembling finished. Used k-mer sizes: 21, 33, 55 \n",
      "\n",
      " * Corrected reads are in /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/corrected/\n",
      " * Assembled contigs are in /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/contigs.fasta\n",
      " * Assembled scaffolds are in /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/scaffolds.fasta\n",
      " * Assembly graph is in /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/assembly_graph.fastg\n",
      " * Assembly graph in GFA format is in /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/assembly_graph_with_scaffolds.gfa\n",
      " * Paths in the assembly graph corresponding to the contigs are in /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/contigs.paths\n",
      " * Paths in the assembly graph corresponding to the scaffolds are in /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/scaffolds.paths\n",
      "\n",
      "======= SPAdes pipeline finished.\n",
      "\n",
      "SPAdes log can be found here: /Users/gregory.wood/Downloads/jupyter_notebooks/lab4/spades_output/spades.log\n",
      "\n",
      "Thank you for using SPAdes!\n"
     ]
    }
   ],
   "source": [
    "! spades.py -1 lab4/read_1.fastq.gz -2 lab4/read_2.fastq.gz -o lab4/spades_output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's all it takes! As you can see, SPAdes has a fairly simple command line interface. \n",
    "\n",
    "#### Use the output from the above command to answer the following questions:\n",
    "1. Take a look at the [SPAdes documentation](https://github.com/ablab/spades); do you see any program options that would let us include a reference genome, if we suspected it might be related to our reads?\n",
    "2. Open the log file found at `lab4/spades_output/spades.log`. This log file contains all the program output from our assembly. What is the kmer size that was used by SPAdes to create the final assembly?\n",
    "3. Our final assembly is contained in `scaffolds.fasta`. What size is this file, compared to `reference.fasta`, the actual reference genome? Based on the coverage we calculated earlier, is this about right? \n",
    "4. Align `scaffolds.fasta` to `reference.fasta` and view it in Jalview. Do you see any obvious places where our assembly is wrong compared to the reference?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exercise 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part B: Homework"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Question 1: The main objective of this homework will be to build your own basic assembly pipeline using the tools we've discussed in this lab. You'll be assembling some reads from an unknown organism using de novo assembly, and then answering some questions about the organism based on the result. Your program can be written in an external shell script, or in Jupyter, but must complete the following steps:\n",
    "1. Perform QC and trimming on the raw FastQ reads, discarding any reads that fail to meet a quality threadhold of N=20\n",
    "2. Calculate the coverage of your resulting trimmed reads. If the coverage is less than 40x, return an error\n",
    "3. Assemble the trimmed reads using SPAdes\n",
    "4. Clean up any temporary files, but make sure to keep the SPAdes scaffold (`scaffolds.fasta`), the fastp/fastqc quality report, and the assembly graph (`assembly_graph.fastg`)\n",
    "\n",
    "If you choose to write your script externally, you should still run it in the cell below. Make sure all results are placed into the lab4/results/ folder at the end of your code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Question 2: Pipelines are often iterative, meaning that we may need to perform multiple runs after changing some parameters. Open the FastP or FastQC report that you generated above. Are the reads that you used for steps 2-4 high-quality? Comment on any steps you would consider taking to refine the reads further in the cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Question 3: Run BLAST on either the command line or using the Web UI to search for the closest match to your assembled `scaffolds.fasta` file. Use the appropriate NCBI databases and download the top hit. Answer the following questions using this data:\n",
    "* What types of hits showed up in your BLAST search? Were all of the organisms from the same species?\n",
    "* How good was the top hit that you downloaded? If you think the match was good, describe why. If not, describe what you think went wrong- does a better hit exist in a different database? Was the assembly bad?\n",
    "* Perform a multiple alignment of your scaffold file to the downloaded sequence. Describe the results- does the alignment match up with your answer to the previous question?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Question 4: As we discussed earlier, de novo assembly is generally difficult to evaluate. One way of telling how \"good\" our assembly is, is checking the De Bruijn graph. If our assembly is good, it will have a single contig, and the graph will be a fully connected \"circle\". If not, there will be (potentially many) disconnected or discontiguous loops. (*Note: a bad assembly can also have a \"good\"  graph- but a good assembly will never have a bad graph!*)\n",
    "\n",
    "We'll use the `bandage` tool to view our graph. Go to the [Bandage GitHub page](https://rrwick.github.io/Bandage/) and download/install the tool. Once it is installed, run it and open up the `assembly_graph.fastg` file. Describe the graph you see below; is it contiguous? Are there extra \"pieces\" of the genome outside the main graph? Based on your knowledge, is our assembly of good quality?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question 4"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:root] *",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
