{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "071d3532",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /root/miniconda3/envs/prokka_env\n",
      "\n",
      "  added / updated specs:\n",
      "    - prokka\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge\n",
      "    _openmp_mutex-4.5          |            1_gnu          22 KB  conda-forge\n",
      "    aragorn-1.2.38             |       hec16e2b_5         134 KB  bioconda\n",
      "    bedtools-2.30.0            |       h468198e_3        15.5 MB  bioconda\n",
      "    blast-2.12.0               |       hf3cf87c_4        22.2 MB  bioconda\n",
      "    c-ares-1.18.1              |       h7f98852_0         113 KB  conda-forge\n",
      "    clustalw-2.1               |       h9f5acd7_7         311 KB  bioconda\n",
      "    curl-7.82.0                |       h7bff187_0          90 KB  conda-forge\n",
      "    expat-2.4.8                |       h27087fc_0         187 KB  conda-forge\n",
      "    hmmer-3.3.2                |       h87f3376_2         9.6 MB  bioconda\n",
      "    infernal-1.1.4             | pl5321hec16e2b_1         3.4 MB  bioconda\n",
      "    keyutils-1.6.1             |       h166bdaf_0         115 KB  conda-forge\n",
      "    krb5-1.19.3                |       h3790be6_0         1.4 MB  conda-forge\n",
      "    ld_impl_linux-64-2.36.1    |       hea4e1c9_2         667 KB  conda-forge\n",
      "    libcurl-7.82.0             |       h7bff187_0         341 KB  conda-forge\n",
      "    libdb-6.2.32               |       h9c3ff4c_0        23.3 MB  conda-forge\n",
      "    libffi-3.4.2               |       h7f98852_5          57 KB  conda-forge\n",
      "    libgcc-ng-11.2.0           |      h1d223b6_15         911 KB  conda-forge\n",
      "    libgomp-11.2.0             |      h1d223b6_15         432 KB  conda-forge\n",
      "    libiconv-1.16              |       h516909a_0         1.4 MB  conda-forge\n",
      "    libidn11-1.34              |       h1cef754_0         227 KB  conda-forge\n",
      "    libnghttp2-1.47.0          |       h727a467_0         808 KB  conda-forge\n",
      "    libnsl-2.0.0               |       h7f98852_0          31 KB  conda-forge\n",
      "    libssh2-1.10.0             |       ha56f1ee_2         233 KB  conda-forge\n",
      "    libstdcxx-ng-11.2.0        |      he4da1e4_15         4.2 MB  conda-forge\n",
      "    libzlib-1.2.11             |    h166bdaf_1014          60 KB  conda-forge\n",
      "    ncurses-6.3                |       h27087fc_1        1002 KB  conda-forge\n",
      "    openjdk-8.0.312            |       h7f98852_0        97.6 MB  conda-forge\n",
      "    openssl-1.1.1n             |       h166bdaf_0         2.1 MB  conda-forge\n",
      "    paml-4.9                   |       hec16e2b_7         994 KB  bioconda\n",
      "    parallel-20220222          |       ha770c72_0         1.9 MB  conda-forge\n",
      "    pcre-8.45                  |       h9c3ff4c_0         253 KB  conda-forge\n",
      "    perl-5.32.1                | 2_h7f98852_perl5        14.4 MB  conda-forge\n",
      "    perl-algorithm-diff-1.201  | pl5321hd8ed1ab_0          32 KB  conda-forge\n",
      "    perl-archive-tar-2.40      | pl5321hdfd78af_0          33 KB  bioconda\n",
      "    perl-base-2.23             | pl5321hdfd78af_2          12 KB  bioconda\n",
      "    perl-bio-asn1-entrezgene-1.73| pl5321hdfd78af_3          24 KB  bioconda\n",
      "    perl-bio-coordinate-1.007001| pl5321hdfd78af_3          36 KB  bioconda\n",
      "    perl-bio-featureio-1.6.905 | pl5321hdfd78af_4          41 KB  bioconda\n",
      "    perl-bio-samtools-1.43     | pl5321h7132678_3         149 KB  bioconda\n",
      "    perl-bio-searchio-hmmer-1.7.3| pl5321hdfd78af_0          58 KB  bioconda\n",
      "    perl-bio-tools-phylo-paml-1.7.3| pl5321hdfd78af_3          58 KB  bioconda\n",
      "    perl-bio-tools-run-alignment-clustalw-1.7.4| pl5321hdfd78af_3          25 KB  bioconda\n",
      "    perl-bio-tools-run-alignment-tcoffee-1.7.4| pl5321hdfd78af_4          25 KB  bioconda\n",
      "    perl-bioperl-1.7.8         |       hdfd78af_1           6 KB  bioconda\n",
      "    perl-bioperl-core-1.7.8    | pl5321hdfd78af_1         1.8 MB  bioconda\n",
      "    perl-bioperl-run-1.007003  | pl5321hdfd78af_0         340 KB  bioconda\n",
      "    perl-business-isbn-3.007   | pl5321hdfd78af_0          18 KB  bioconda\n",
      "    perl-business-isbn-data-20210112.006| pl5321hdfd78af_0          21 KB  bioconda\n",
      "    perl-capture-tiny-0.48     | pl5321ha770c72_1          16 KB  conda-forge\n",
      "    perl-carp-1.50             | pl5321hd8ed1ab_0          22 KB  conda-forge\n",
      "    perl-class-data-inheritable-0.09| pl5321hdfd78af_0          10 KB  bioconda\n",
      "    perl-common-sense-3.75     | pl5321hdfd78af_0          13 KB  bioconda\n",
      "    perl-compress-raw-bzip2-2.103| pl5321h87f3376_0          47 KB  bioconda\n",
      "    perl-compress-raw-zlib-2.103| pl5321h87f3376_0          75 KB  bioconda\n",
      "    perl-constant-1.33         | pl5321hd8ed1ab_0          15 KB  conda-forge\n",
      "    perl-data-dumper-2.183     | pl5321hec16e2b_1          38 KB  bioconda\n",
      "    perl-db_file-1.857         | pl5321h166bdaf_0          59 KB  conda-forge\n",
      "    perl-devel-stacktrace-2.04 | pl5321hdfd78af_1          17 KB  bioconda\n",
      "    perl-digest-hmac-1.04      | pl5321hdfd78af_0          10 KB  bioconda\n",
      "    perl-digest-md5-2.58       | pl5321hec16e2b_1          19 KB  bioconda\n",
      "    perl-encode-3.17           | pl5321hec16e2b_0         2.1 MB  bioconda\n",
      "    perl-encode-locale-1.05    | pl5321hdfd78af_7          13 KB  bioconda\n",
      "    perl-exception-class-1.45  | pl5321hdfd78af_0          21 KB  bioconda\n",
      "    perl-exporter-5.74         | pl5321hd8ed1ab_0          19 KB  conda-forge\n",
      "    perl-exporter-tiny-1.002002| pl5321hdfd78af_0          25 KB  bioconda\n",
      "    perl-extutils-makemaker-7.64| pl5321hd8ed1ab_0         145 KB  conda-forge\n",
      "    perl-file-listing-6.14     | pl5321hdfd78af_0          14 KB  bioconda\n",
      "    perl-file-path-2.18        | pl5321hd8ed1ab_0          22 KB  conda-forge\n",
      "    perl-file-slurp-tiny-0.004 | pl5321hdfd78af_2          10 KB  bioconda\n",
      "    perl-file-sort-1.01        | pl5321hdfd78af_3          20 KB  bioconda\n",
      "    perl-file-spec-3.48_01     | pl5321hdfd78af_2           6 KB  bioconda\n",
      "    perl-file-temp-0.2304      | pl5321hd8ed1ab_0          31 KB  conda-forge\n",
      "    perl-getopt-long-2.52      | pl5321hdfd78af_0          30 KB  bioconda\n",
      "    perl-html-parser-3.72      | pl5321h9f5acd7_8          66 KB  bioconda\n",
      "    perl-html-tagset-3.20      | pl5321hdfd78af_4          14 KB  bioconda\n",
      "    perl-http-cookies-6.10     | pl5321hdfd78af_0          22 KB  bioconda\n",
      "    perl-http-daemon-6.14      | pl5321hdfd78af_0          20 KB  bioconda\n",
      "    perl-http-date-6.05        | pl5321hdfd78af_0          13 KB  bioconda\n",
      "    perl-http-message-6.36     | pl5321hdfd78af_0          54 KB  bioconda\n",
      "    perl-http-negotiate-6.01   | pl5321hdfd78af_4          14 KB  bioconda\n",
      "    perl-io-compress-2.102     | pl5321h87f3376_1          83 KB  bioconda\n",
      "    perl-io-html-1.004         | pl5321hdfd78af_0          16 KB  bioconda\n",
      "    perl-io-socket-ssl-2.074   | pl5321hdfd78af_0         162 KB  bioconda\n",
      "    perl-io-string-1.08        | pl5321hdfd78af_4          12 KB  bioconda\n",
      "    perl-io-tty-1.16           | pl5321hec16e2b_1          32 KB  bioconda\n",
      "    perl-io-zlib-1.11          | pl5321hdfd78af_0          11 KB  bioconda\n",
      "    perl-ipc-run-20200505.0    | pl5321hdfd78af_0          76 KB  bioconda\n",
      "    perl-json-4.05             | pl5321hdfd78af_0          54 KB  bioconda\n",
      "    perl-json-xs-2.34          | pl5321h9f5acd7_5          64 KB  bioconda\n",
      "    perl-libwww-perl-6.62      | pl5321hdfd78af_0          97 KB  bioconda\n",
      "    perl-libxml-perl-0.08      | pl5321hdfd78af_3          40 KB  bioconda\n",
      "    perl-list-moreutils-0.430  | pl5321hdfd78af_0          32 KB  bioconda\n",
      "    perl-list-moreutils-xs-0.430| pl5321hec16e2b_1          49 KB  bioconda\n",
      "    perl-lwp-mediatypes-6.04   | pl5321hdfd78af_1          24 KB  bioconda\n",
      "    perl-mime-base64-3.16      | pl5321hec16e2b_2          17 KB  bioconda\n",
      "    perl-net-http-6.22         | pl5321hdfd78af_0          21 KB  bioconda\n",
      "    perl-net-ssleay-1.92       | pl5321h0e0aaa8_1         302 KB  bioconda\n",
      "    perl-ntlm-1.09             | pl5321hdfd78af_5          16 KB  bioconda\n",
      "    perl-parent-0.238          | pl5321hd8ed1ab_0          11 KB  conda-forge\n",
      "    perl-pathtools-3.75        | pl5321hec16e2b_3          42 KB  bioconda\n",
      "    perl-scalar-list-utils-1.62| pl5321hec16e2b_0          43 KB  bioconda\n",
      "    perl-socket-2.027          | pl5321hec16e2b_3          33 KB  bioconda\n",
      "    perl-sub-uplevel-0.2800    | pl5321hec16e2b_4          14 KB  bioconda\n",
      "    perl-test-deep-1.130       | pl5321hdfd78af_0          37 KB  bioconda\n",
      "    perl-test-differences-0.69 | pl5321hdfd78af_0          19 KB  bioconda\n",
      "    perl-test-exception-0.43   | pl5321hdfd78af_3          15 KB  bioconda\n",
      "    perl-test-most-0.37        | pl5321hdfd78af_0          20 KB  bioconda\n",
      "    perl-test-warn-0.36        | pl5321hdfd78af_2          14 KB  bioconda\n",
      "    perl-text-diff-1.45        | pl5321hdfd78af_1          23 KB  bioconda\n",
      "    perl-time-local-1.30       | pl5321hdfd78af_0          12 KB  bioconda\n",
      "    perl-timedate-2.33         | pl5321hdfd78af_2          31 KB  bioconda\n",
      "    perl-tree-dag_node-1.32    | pl5321hdfd78af_0          43 KB  bioconda\n",
      "    perl-try-tiny-0.31         | pl5321hdfd78af_0          18 KB  bioconda\n",
      "    perl-types-serialiser-1.01 | pl5321hdfd78af_0          13 KB  bioconda\n",
      "    perl-uri-5.10              | pl5321hdfd78af_0          57 KB  bioconda\n",
      "    perl-url-encode-0.03       | pl5321h9ee0642_0          15 KB  bioconda\n",
      "    perl-www-robotrules-6.02   | pl5321hdfd78af_4          14 KB  bioconda\n",
      "    perl-xml-dom-1.46          | pl5321hdfd78af_1          59 KB  bioconda\n",
      "    perl-xml-dom-xpath-0.14    | pl5321hdfd78af_2          12 KB  bioconda\n",
      "    perl-xml-namespacesupport-1.12| pl5321hdfd78af_1          14 KB  bioconda\n",
      "    perl-xml-parser-2.44_01    |pl5321hc3e0081_1003         164 KB  conda-forge\n",
      "    perl-xml-regexp-0.04       | pl5321hdfd78af_3          10 KB  bioconda\n",
      "    perl-xml-sax-1.02          | pl5321hdfd78af_1          37 KB  bioconda\n",
      "    perl-xml-sax-base-1.09     | pl5321hdfd78af_1          21 KB  bioconda\n",
      "    perl-xml-sax-expat-0.51    | pl5321hdfd78af_4          12 KB  bioconda\n",
      "    perl-xml-simple-2.25       | pl5321hdfd78af_2          49 KB  bioconda\n",
      "    perl-xml-xpathengine-0.14  | pl5321hdfd78af_3          28 KB  bioconda\n",
      "    pip-22.0.4                 |     pyhd8ed1ab_0         1.5 MB  conda-forge\n",
      "    prodigal-2.6.3             |       hec16e2b_4         777 KB  bioconda\n",
      "    prokka-1.14.6              | pl5321hdfd78af_4       212.7 MB  bioconda\n",
      "    python-3.7.12              |hb7a2778_100_cpython        57.3 MB  conda-forge\n",
      "    readline-8.1               |       h46c0cb4_0         295 KB  conda-forge\n",
      "    setuptools-62.0.0          |   py37h89c1867_0         1.2 MB  conda-forge\n",
      "    sqlite-3.37.1              |       h4ff8645_0         1.5 MB  conda-forge\n",
      "    tbl2asn-forever-25.7.2f    |       hec16e2b_2         4.3 MB  bioconda\n",
      "    tk-8.6.12                  |       h27826a3_0         3.3 MB  conda-forge\n",
      "    wheel-0.37.1               |     pyhd8ed1ab_0          31 KB  conda-forge\n",
      "    xz-5.2.5                   |       h516909a_1         343 KB  conda-forge\n",
      "    zlib-1.2.11                |    h166bdaf_1014          88 KB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:       494.8 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
      "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
      "  aragorn            bioconda/linux-64::aragorn-1.2.38-hec16e2b_5\n",
      "  barrnap            bioconda/noarch::barrnap-0.9-hdfd78af_4\n",
      "  bedtools           bioconda/linux-64::bedtools-2.30.0-h468198e_3\n",
      "  blast              bioconda/linux-64::blast-2.12.0-hf3cf87c_4\n",
      "  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4\n",
      "  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0\n",
      "  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0\n",
      "  clustalw           bioconda/linux-64::clustalw-2.1-h9f5acd7_7\n",
      "  curl               conda-forge/linux-64::curl-7.82.0-h7bff187_0\n",
      "  entrez-direct      bioconda/linux-64::entrez-direct-16.2-he881be0_0\n",
      "  expat              conda-forge/linux-64::expat-2.4.8-h27087fc_0\n",
      "  hmmer              bioconda/linux-64::hmmer-3.3.2-h87f3376_2\n",
      "  infernal           bioconda/linux-64::infernal-1.1.4-pl5321hec16e2b_1\n",
      "  keyutils           conda-forge/linux-64::keyutils-1.6.1-h166bdaf_0\n",
      "  krb5               conda-forge/linux-64::krb5-1.19.3-h3790be6_0\n",
      "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.36.1-hea4e1c9_2\n",
      "  libcurl            conda-forge/linux-64::libcurl-7.82.0-h7bff187_0\n",
      "  libdb              conda-forge/linux-64::libdb-6.2.32-h9c3ff4c_0\n",
      "  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2\n",
      "  libev              conda-forge/linux-64::libev-4.33-h516909a_1\n",
      "  libffi             conda-forge/linux-64::libffi-3.4.2-h7f98852_5\n",
      "  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_15\n",
      "  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_15\n",
      "  libiconv           conda-forge/linux-64::libiconv-1.16-h516909a_0\n",
      "  libidn11           conda-forge/linux-64::libidn11-1.34-h1cef754_0\n",
      "  libnghttp2         conda-forge/linux-64::libnghttp2-1.47.0-h727a467_0\n",
      "  libnsl             conda-forge/linux-64::libnsl-2.0.0-h7f98852_0\n",
      "  libssh2            conda-forge/linux-64::libssh2-1.10.0-ha56f1ee_2\n",
      "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_15\n",
      "  libzlib            conda-forge/linux-64::libzlib-1.2.11-h166bdaf_1014\n",
      "  minced             bioconda/noarch::minced-0.4.2-hdfd78af_1\n",
      "  ncurses            conda-forge/linux-64::ncurses-6.3-h27087fc_1\n",
      "  openjdk            conda-forge/linux-64::openjdk-8.0.312-h7f98852_0\n",
      "  openssl            conda-forge/linux-64::openssl-1.1.1n-h166bdaf_0\n",
      "  paml               bioconda/linux-64::paml-4.9-hec16e2b_7\n",
      "  parallel           conda-forge/linux-64::parallel-20220222-ha770c72_0\n",
      "  pcre               conda-forge/linux-64::pcre-8.45-h9c3ff4c_0\n",
      "  perl               conda-forge/linux-64::perl-5.32.1-2_h7f98852_perl5\n",
      "  perl-algorithm-di~ conda-forge/noarch::perl-algorithm-diff-1.201-pl5321hd8ed1ab_0\n",
      "  perl-archive-tar   bioconda/noarch::perl-archive-tar-2.40-pl5321hdfd78af_0\n",
      "  perl-base          bioconda/noarch::perl-base-2.23-pl5321hdfd78af_2\n",
      "  perl-bio-asn1-ent~ bioconda/noarch::perl-bio-asn1-entrezgene-1.73-pl5321hdfd78af_3\n",
      "  perl-bio-coordina~ bioconda/noarch::perl-bio-coordinate-1.007001-pl5321hdfd78af_3\n",
      "  perl-bio-featureio bioconda/noarch::perl-bio-featureio-1.6.905-pl5321hdfd78af_4\n",
      "  perl-bio-samtools  bioconda/linux-64::perl-bio-samtools-1.43-pl5321h7132678_3\n",
      "  perl-bio-searchio~ bioconda/noarch::perl-bio-searchio-hmmer-1.7.3-pl5321hdfd78af_0\n",
      "  perl-bio-tools-ph~ bioconda/noarch::perl-bio-tools-phylo-paml-1.7.3-pl5321hdfd78af_3\n",
      "  perl-bio-tools-ru~ bioconda/noarch::perl-bio-tools-run-alignment-clustalw-1.7.4-pl5321hdfd78af_3\n",
      "  perl-bio-tools-ru~ bioconda/noarch::perl-bio-tools-run-alignment-tcoffee-1.7.4-pl5321hdfd78af_4\n",
      "  perl-bioperl       bioconda/noarch::perl-bioperl-1.7.8-hdfd78af_1\n",
      "  perl-bioperl-core  bioconda/noarch::perl-bioperl-core-1.7.8-pl5321hdfd78af_1\n",
      "  perl-bioperl-run   bioconda/noarch::perl-bioperl-run-1.007003-pl5321hdfd78af_0\n",
      "  perl-business-isbn bioconda/noarch::perl-business-isbn-3.007-pl5321hdfd78af_0\n",
      "  perl-business-isb~ bioconda/noarch::perl-business-isbn-data-20210112.006-pl5321hdfd78af_0\n",
      "  perl-capture-tiny  conda-forge/linux-64::perl-capture-tiny-0.48-pl5321ha770c72_1\n",
      "  perl-carp          conda-forge/noarch::perl-carp-1.50-pl5321hd8ed1ab_0\n",
      "  perl-class-data-i~ bioconda/noarch::perl-class-data-inheritable-0.09-pl5321hdfd78af_0\n",
      "  perl-common-sense  bioconda/noarch::perl-common-sense-3.75-pl5321hdfd78af_0\n",
      "  perl-compress-raw~ bioconda/linux-64::perl-compress-raw-bzip2-2.103-pl5321h87f3376_0\n",
      "  perl-compress-raw~ bioconda/linux-64::perl-compress-raw-zlib-2.103-pl5321h87f3376_0\n",
      "  perl-constant      conda-forge/noarch::perl-constant-1.33-pl5321hd8ed1ab_0\n",
      "  perl-data-dumper   bioconda/linux-64::perl-data-dumper-2.183-pl5321hec16e2b_1\n",
      "  perl-db_file       conda-forge/linux-64::perl-db_file-1.857-pl5321h166bdaf_0\n",
      "  perl-devel-stackt~ bioconda/noarch::perl-devel-stacktrace-2.04-pl5321hdfd78af_1\n",
      "  perl-digest-hmac   bioconda/noarch::perl-digest-hmac-1.04-pl5321hdfd78af_0\n",
      "  perl-digest-md5    bioconda/linux-64::perl-digest-md5-2.58-pl5321hec16e2b_1\n",
      "  perl-encode        bioconda/linux-64::perl-encode-3.17-pl5321hec16e2b_0\n",
      "  perl-encode-locale bioconda/noarch::perl-encode-locale-1.05-pl5321hdfd78af_7\n",
      "  perl-exception-cl~ bioconda/noarch::perl-exception-class-1.45-pl5321hdfd78af_0\n",
      "  perl-exporter      conda-forge/noarch::perl-exporter-5.74-pl5321hd8ed1ab_0\n",
      "  perl-exporter-tiny bioconda/noarch::perl-exporter-tiny-1.002002-pl5321hdfd78af_0\n",
      "  perl-extutils-mak~ conda-forge/noarch::perl-extutils-makemaker-7.64-pl5321hd8ed1ab_0\n",
      "  perl-file-listing  bioconda/noarch::perl-file-listing-6.14-pl5321hdfd78af_0\n",
      "  perl-file-path     conda-forge/noarch::perl-file-path-2.18-pl5321hd8ed1ab_0\n",
      "  perl-file-slurp-t~ bioconda/noarch::perl-file-slurp-tiny-0.004-pl5321hdfd78af_2\n",
      "  perl-file-sort     bioconda/noarch::perl-file-sort-1.01-pl5321hdfd78af_3\n",
      "  perl-file-spec     bioconda/noarch::perl-file-spec-3.48_01-pl5321hdfd78af_2\n",
      "  perl-file-temp     conda-forge/noarch::perl-file-temp-0.2304-pl5321hd8ed1ab_0\n",
      "  perl-getopt-long   bioconda/noarch::perl-getopt-long-2.52-pl5321hdfd78af_0\n",
      "  perl-html-parser   bioconda/linux-64::perl-html-parser-3.72-pl5321h9f5acd7_8\n",
      "  perl-html-tagset   bioconda/noarch::perl-html-tagset-3.20-pl5321hdfd78af_4\n",
      "  perl-http-cookies  bioconda/noarch::perl-http-cookies-6.10-pl5321hdfd78af_0\n",
      "  perl-http-daemon   bioconda/noarch::perl-http-daemon-6.14-pl5321hdfd78af_0\n",
      "  perl-http-date     bioconda/noarch::perl-http-date-6.05-pl5321hdfd78af_0\n",
      "  perl-http-message  bioconda/noarch::perl-http-message-6.36-pl5321hdfd78af_0\n",
      "  perl-http-negotia~ bioconda/noarch::perl-http-negotiate-6.01-pl5321hdfd78af_4\n",
      "  perl-io-compress   bioconda/linux-64::perl-io-compress-2.102-pl5321h87f3376_1\n",
      "  perl-io-html       bioconda/noarch::perl-io-html-1.004-pl5321hdfd78af_0\n",
      "  perl-io-socket-ssl bioconda/noarch::perl-io-socket-ssl-2.074-pl5321hdfd78af_0\n",
      "  perl-io-string     bioconda/noarch::perl-io-string-1.08-pl5321hdfd78af_4\n",
      "  perl-io-tty        bioconda/linux-64::perl-io-tty-1.16-pl5321hec16e2b_1\n",
      "  perl-io-zlib       bioconda/noarch::perl-io-zlib-1.11-pl5321hdfd78af_0\n",
      "  perl-ipc-run       bioconda/noarch::perl-ipc-run-20200505.0-pl5321hdfd78af_0\n",
      "  perl-json          bioconda/noarch::perl-json-4.05-pl5321hdfd78af_0\n",
      "  perl-json-xs       bioconda/linux-64::perl-json-xs-2.34-pl5321h9f5acd7_5\n",
      "  perl-libwww-perl   bioconda/noarch::perl-libwww-perl-6.62-pl5321hdfd78af_0\n",
      "  perl-libxml-perl   bioconda/noarch::perl-libxml-perl-0.08-pl5321hdfd78af_3\n",
      "  perl-list-moreuti~ bioconda/noarch::perl-list-moreutils-0.430-pl5321hdfd78af_0\n",
      "  perl-list-moreuti~ bioconda/linux-64::perl-list-moreutils-xs-0.430-pl5321hec16e2b_1\n",
      "  perl-lwp-mediatyp~ bioconda/noarch::perl-lwp-mediatypes-6.04-pl5321hdfd78af_1\n",
      "  perl-mime-base64   bioconda/linux-64::perl-mime-base64-3.16-pl5321hec16e2b_2\n",
      "  perl-net-http      bioconda/noarch::perl-net-http-6.22-pl5321hdfd78af_0\n",
      "  perl-net-ssleay    bioconda/linux-64::perl-net-ssleay-1.92-pl5321h0e0aaa8_1\n",
      "  perl-ntlm          bioconda/noarch::perl-ntlm-1.09-pl5321hdfd78af_5\n",
      "  perl-parent        conda-forge/noarch::perl-parent-0.238-pl5321hd8ed1ab_0\n",
      "  perl-pathtools     bioconda/linux-64::perl-pathtools-3.75-pl5321hec16e2b_3\n",
      "  perl-scalar-list-~ bioconda/linux-64::perl-scalar-list-utils-1.62-pl5321hec16e2b_0\n",
      "  perl-socket        bioconda/linux-64::perl-socket-2.027-pl5321hec16e2b_3\n",
      "  perl-sub-uplevel   bioconda/linux-64::perl-sub-uplevel-0.2800-pl5321hec16e2b_4\n",
      "  perl-test-deep     bioconda/noarch::perl-test-deep-1.130-pl5321hdfd78af_0\n",
      "  perl-test-differe~ bioconda/noarch::perl-test-differences-0.69-pl5321hdfd78af_0\n",
      "  perl-test-excepti~ bioconda/noarch::perl-test-exception-0.43-pl5321hdfd78af_3\n",
      "  perl-test-most     bioconda/noarch::perl-test-most-0.37-pl5321hdfd78af_0\n",
      "  perl-test-warn     bioconda/noarch::perl-test-warn-0.36-pl5321hdfd78af_2\n",
      "  perl-text-diff     bioconda/noarch::perl-text-diff-1.45-pl5321hdfd78af_1\n",
      "  perl-time-local    bioconda/noarch::perl-time-local-1.30-pl5321hdfd78af_0\n",
      "  perl-timedate      bioconda/noarch::perl-timedate-2.33-pl5321hdfd78af_2\n",
      "  perl-tree-dag_node bioconda/noarch::perl-tree-dag_node-1.32-pl5321hdfd78af_0\n",
      "  perl-try-tiny      bioconda/noarch::perl-try-tiny-0.31-pl5321hdfd78af_0\n",
      "  perl-types-serial~ bioconda/noarch::perl-types-serialiser-1.01-pl5321hdfd78af_0\n",
      "  perl-uri           bioconda/noarch::perl-uri-5.10-pl5321hdfd78af_0\n",
      "  perl-url-encode    bioconda/linux-64::perl-url-encode-0.03-pl5321h9ee0642_0\n",
      "  perl-www-robotrul~ bioconda/noarch::perl-www-robotrules-6.02-pl5321hdfd78af_4\n",
      "  perl-xml-dom       bioconda/noarch::perl-xml-dom-1.46-pl5321hdfd78af_1\n",
      "  perl-xml-dom-xpath bioconda/noarch::perl-xml-dom-xpath-0.14-pl5321hdfd78af_2\n",
      "  perl-xml-namespac~ bioconda/noarch::perl-xml-namespacesupport-1.12-pl5321hdfd78af_1\n",
      "  perl-xml-parser    conda-forge/linux-64::perl-xml-parser-2.44_01-pl5321hc3e0081_1003\n",
      "  perl-xml-regexp    bioconda/noarch::perl-xml-regexp-0.04-pl5321hdfd78af_3\n",
      "  perl-xml-sax       bioconda/noarch::perl-xml-sax-1.02-pl5321hdfd78af_1\n",
      "  perl-xml-sax-base  bioconda/noarch::perl-xml-sax-base-1.09-pl5321hdfd78af_1\n",
      "  perl-xml-sax-expat bioconda/noarch::perl-xml-sax-expat-0.51-pl5321hdfd78af_4\n",
      "  perl-xml-simple    bioconda/noarch::perl-xml-simple-2.25-pl5321hdfd78af_2\n",
      "  perl-xml-xpatheng~ bioconda/noarch::perl-xml-xpathengine-0.14-pl5321hdfd78af_3\n",
      "  pip                conda-forge/noarch::pip-22.0.4-pyhd8ed1ab_0\n",
      "  prodigal           bioconda/linux-64::prodigal-2.6.3-hec16e2b_4\n",
      "  prokka             bioconda/noarch::prokka-1.14.6-pl5321hdfd78af_4\n",
      "  python             conda-forge/linux-64::python-3.7.12-hb7a2778_100_cpython\n",
      "  python_abi         conda-forge/linux-64::python_abi-3.7-2_cp37m\n",
      "  readline           conda-forge/linux-64::readline-8.1-h46c0cb4_0\n",
      "  setuptools         conda-forge/linux-64::setuptools-62.0.0-py37h89c1867_0\n",
      "  sqlite             conda-forge/linux-64::sqlite-3.37.1-h4ff8645_0\n",
      "  t_coffee           bioconda/linux-64::t_coffee-11.0.8-py37hea885bf_8\n",
      "  tbl2asn-forever    bioconda/linux-64::tbl2asn-forever-25.7.2f-hec16e2b_2\n",
      "  tk                 conda-forge/linux-64::tk-8.6.12-h27826a3_0\n",
      "  wheel              conda-forge/noarch::wheel-0.37.1-pyhd8ed1ab_0\n",
      "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
      "  zlib               conda-forge/linux-64::zlib-1.2.11-h166bdaf_1014\n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "perl-json-xs-2.34    | 64 KB     | ########## | 100% \n",
      "perl-text-diff-1.45  | 23 KB     | ########## | 100% \n",
      "perl-mime-base64-3.1 | 17 KB     | ########## | 100% \n",
      "perl-http-cookies-6. | 22 KB     | ########## | 100% \n",
      "perl-digest-md5-2.58 | 19 KB     | ########## | 100% \n",
      "c-ares-1.18.1        | 113 KB    | ########## | 100% \n",
      "libstdcxx-ng-11.2.0  | 4.2 MB    | ########## | 100% \n",
      "libgomp-11.2.0       | 432 KB    | ########## | 100% \n",
      "libidn11-1.34        | 227 KB    | ########## | 100% \n",
      "perl-algorithm-diff- | 32 KB     | ########## | 100% \n",
      "perl-lwp-mediatypes- | 24 KB     | ########## | 100% \n",
      "perl-bio-asn1-entrez | 24 KB     | ########## | 100% \n",
      "perl-xml-dom-xpath-0 | 12 KB     | ########## | 100% \n",
      "perl-data-dumper-2.1 | 38 KB     | ########## | 100% \n",
      "bedtools-2.30.0      | 15.5 MB   | ########## | 100% \n",
      "perl-sub-uplevel-0.2 | 14 KB     | ########## | 100% \n",
      "perl-ipc-run-2020050 | 76 KB     | ########## | 100% \n",
      "keyutils-1.6.1       | 115 KB    | ########## | 100% \n",
      "libiconv-1.16        | 1.4 MB    | ########## | 100% \n",
      "pip-22.0.4           | 1.5 MB    | ########## | 100% \n",
      "infernal-1.1.4       | 3.4 MB    | ########## | 100% \n",
      "perl-bio-tools-phylo | 58 KB     | ########## | 100% \n",
      "setuptools-62.0.0    | 1.2 MB    | ########## | 100% \n",
      "perl-xml-namespacesu | 14 KB     | ########## | 100% \n",
      "perl-carp-1.50       | 22 KB     | ########## | 100% \n",
      "perl-archive-tar-2.4 | 33 KB     | ########## | 100% \n",
      "aragorn-1.2.38       | 134 KB    | ########## | 100% \n",
      "perl-net-ssleay-1.92 | 302 KB    | ########## | 100% \n",
      "perl-file-path-2.18  | 22 KB     | ########## | 100% \n",
      "perl-xml-xpathengine | 28 KB     | ########## | 100% \n",
      "curl-7.82.0          | 90 KB     | ########## | 100% \n",
      "perl-html-parser-3.7 | 66 KB     | ########## | 100% \n",
      "perl-extutils-makema | 145 KB    | ########## | 100% \n",
      "perl-bio-tools-run-a | 25 KB     | ########## | 100% \n",
      "perl-constant-1.33   | 15 KB     | ########## | 100% \n",
      "perl-xml-dom-1.46    | 59 KB     | ########## | 100% \n",
      "perl-file-listing-6. | 14 KB     | ########## | 100% \n",
      "perl-io-zlib-1.11    | 11 KB     | ########## | 100% \n",
      "perl-list-moreutils- | 32 KB     | ########## | 100% \n",
      "perl-bio-samtools-1. | 149 KB    | ########## | 100% \n",
      "perl-http-message-6. | 54 KB     | ########## | 100% \n",
      "perl-5.32.1          | 14.4 MB   | ########## | 100% \n",
      "perl-ntlm-1.09       | 16 KB     | ########## | 100% \n",
      "perl-io-socket-ssl-2 | 162 KB    | ########## | 100% \n",
      "blast-2.12.0         | 22.2 MB   | ########## | 100% \n",
      "libssh2-1.10.0       | 233 KB    | ########## | 100% \n",
      "perl-io-string-1.08  | 12 KB     | ########## | 100% \n",
      "perl-digest-hmac-1.0 | 10 KB     | ########## | 100% \n",
      "sqlite-3.37.1        | 1.5 MB    | ########## | 100% \n",
      "perl-http-date-6.05  | 13 KB     | ########## | 100% \n",
      "perl-devel-stacktrac | 17 KB     | ########## | 100% \n",
      "readline-8.1         | 295 KB    | ########## | 100% \n",
      "perl-url-encode-0.03 | 15 KB     | ########## | 100% \n",
      "perl-db_file-1.857   | 59 KB     | ########## | 100% \n",
      "_openmp_mutex-4.5    | 22 KB     | ########## | 100% \n",
      "perl-file-spec-3.48_ | 6 KB      | ########## | 100% \n",
      "perl-io-tty-1.16     | 32 KB     | ########## | 100% \n",
      "perl-io-html-1.004   | 16 KB     | ########## | 100% \n",
      "perl-net-http-6.22   | 21 KB     | ########## | 100% \n",
      "xz-5.2.5             | 343 KB    | ########## | 100% \n",
      "perl-encode-locale-1 | 13 KB     | ########## | 100% \n",
      "perl-bioperl-core-1. | 1.8 MB    | ########## | 100% \n",
      "perl-file-temp-0.230 | 31 KB     | ########## | 100% \n",
      "perl-json-4.05       | 54 KB     | ########## | 100% \n",
      "perl-bioperl-1.7.8   | 6 KB      | ########## | 100% \n",
      "perl-xml-sax-base-1. | 21 KB     | ########## | 100% \n",
      "openjdk-8.0.312      | 97.6 MB   | ########## | 100% \n",
      "prodigal-2.6.3       | 777 KB    | ########## | 100% \n",
      "libnghttp2-1.47.0    | 808 KB    | ########## | 100% \n",
      "krb5-1.19.3          | 1.4 MB    | ########## | 100% \n",
      "perl-getopt-long-2.5 | 30 KB     | ########## | 100% \n",
      "perl-test-exception- | 15 KB     | ########## | 100% \n",
      "perl-xml-sax-1.02    | 37 KB     | ########## | 100% \n",
      "perl-test-deep-1.130 | 37 KB     | ########## | 100% \n",
      "clustalw-2.1         | 311 KB    | ########## | 100% \n",
      "prokka-1.14.6        | 212.7 MB  | ########## | 100% \n",
      "wheel-0.37.1         | 31 KB     | ########## | 100% \n",
      "perl-test-most-0.37  | 20 KB     | ########## | 100% \n",
      "perl-www-robotrules- | 14 KB     | ########## | 100% \n",
      "perl-time-local-1.30 | 12 KB     | ########## | 100% \n",
      "perl-http-daemon-6.1 | 20 KB     | ########## | 100% \n",
      "perl-io-compress-2.1 | 83 KB     | ########## | 100% \n",
      "perl-file-slurp-tiny | 10 KB     | ########## | 100% \n",
      "perl-exporter-tiny-1 | 25 KB     | ########## | 100% \n",
      "perl-class-data-inhe | 10 KB     | ########## | 100% \n",
      "openssl-1.1.1n       | 2.1 MB    | ########## | 100% \n",
      "perl-uri-5.10        | 57 KB     | ########## | 100% \n",
      "libgcc-ng-11.2.0     | 911 KB    | ########## | 100% \n",
      "ncurses-6.3          | 1002 KB   | ########## | 100% \n",
      "perl-bio-featureio-1 | 41 KB     | ########## | 100% \n",
      "perl-test-difference | 19 KB     | ########## | 100% \n",
      "perl-common-sense-3. | 13 KB     | ########## | 100% \n",
      "tk-8.6.12            | 3.3 MB    | ########## | 100% \n",
      "libnsl-2.0.0         | 31 KB     | ########## | 100% \n",
      "expat-2.4.8          | 187 KB    | ########## | 100% \n",
      "libffi-3.4.2         | 57 KB     | ########## | 100% \n",
      "perl-bio-tools-run-a | 25 KB     | ########## | 100% \n",
      "perl-bio-searchio-hm | 58 KB     | ########## | 100% \n",
      "libzlib-1.2.11       | 60 KB     | ########## | 100% \n",
      "perl-list-moreutils- | 49 KB     | ########## | 100% \n",
      "perl-scalar-list-uti | 43 KB     | ########## | 100% \n",
      "perl-http-negotiate- | 14 KB     | ########## | 100% \n",
      "python-3.7.12        | 57.3 MB   | ########## | 100% \n",
      "perl-bio-coordinate- | 36 KB     | ########## | 100% \n",
      "perl-libxml-perl-0.0 | 40 KB     | ########## | 100% \n",
      "perl-bioperl-run-1.0 | 340 KB    | ########## | 100% \n",
      "perl-compress-raw-bz | 47 KB     | ########## | 100% \n",
      "perl-parent-0.238    | 11 KB     | ########## | 100% \n",
      "zlib-1.2.11          | 88 KB     | ########## | 100% \n",
      "perl-base-2.23       | 12 KB     | ########## | 100% \n",
      "libdb-6.2.32         | 23.3 MB   | ########## | 100% \n",
      "perl-test-warn-0.36  | 14 KB     | ########## | 100% \n",
      "hmmer-3.3.2          | 9.6 MB    | ########## | 100% \n",
      "perl-xml-parser-2.44 | 164 KB    | ########## | 100% \n",
      "perl-tree-dag_node-1 | 43 KB     | ########## | 100% \n",
      "perl-timedate-2.33   | 31 KB     | ########## | 100% \n",
      "tbl2asn-forever-25.7 | 4.3 MB    | ########## | 100% \n",
      "perl-xml-simple-2.25 | 49 KB     | ########## | 100% \n",
      "perl-html-tagset-3.2 | 14 KB     | ########## | 100% \n",
      "perl-xml-sax-expat-0 | 12 KB     | ########## | 100% \n",
      "_libgcc_mutex-0.1    | 3 KB      | ########## | 100% \n",
      "paml-4.9             | 994 KB    | ########## | 100% \n",
      "pcre-8.45            | 253 KB    | ########## | 100% \n",
      "perl-libwww-perl-6.6 | 97 KB     | ########## | 100% \n",
      "perl-types-serialise | 13 KB     | ########## | 100% \n",
      "parallel-20220222    | 1.9 MB    | ########## | 100% \n",
      "perl-business-isbn-3 | 18 KB     | ########## | 100% \n",
      "perl-encode-3.17     | 2.1 MB    | ########## | 100% \n",
      "perl-capture-tiny-0. | 16 KB     | ########## | 100% \n",
      "libcurl-7.82.0       | 341 KB    | ########## | 100% \n",
      "perl-compress-raw-zl | 75 KB     | ########## | 100% \n",
      "perl-pathtools-3.75  | 42 KB     | ########## | 100% \n",
      "perl-try-tiny-0.31   | 18 KB     | ########## | 100% \n",
      "perl-business-isbn-d | 21 KB     | ########## | 100% \n",
      "perl-socket-2.027    | 33 KB     | ########## | 100% \n",
      "perl-file-sort-1.01  | 20 KB     | ########## | 100% \n",
      "ld_impl_linux-64-2.3 | 667 KB    | ########## | 100% \n",
      "perl-xml-regexp-0.04 | 10 KB     | ########## | 100% \n",
      "perl-exception-class | 21 KB     | ########## | 100% \n",
      "perl-exporter-5.74   | 19 KB     | ########## | 100% \n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate prokka_env\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "Collecting package metadata (current_repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /root/miniconda3/envs/prokka_env\n",
      "\n",
      "  added / updated specs:\n",
      "    - ipykernel\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  backcall           anaconda/noarch::backcall-0.2.0-py_0\n",
      "  decorator          anaconda/noarch::decorator-4.4.2-py_0\n",
      "  ipykernel          anaconda/linux-64::ipykernel-5.3.4-py37h5ca1d4c_0\n",
      "  ipython            anaconda/linux-64::ipython-7.18.1-py37h5ca1d4c_0\n",
      "  ipython_genutils   anaconda/linux-64::ipython_genutils-0.2.0-py37_0\n",
      "  jedi               pkgs/main/linux-64::jedi-0.18.1-py37h06a4308_1\n",
      "  jupyter_client     anaconda/noarch::jupyter_client-6.1.7-py_0\n",
      "  jupyter_core       anaconda/linux-64::jupyter_core-4.6.3-py37_0\n",
      "  libsodium          anaconda/linux-64::libsodium-1.0.18-h7b6447c_0\n",
      "  parso              anaconda/noarch::parso-0.8.0-py_0\n",
      "  pexpect            anaconda/linux-64::pexpect-4.8.0-py37_1\n",
      "  pickleshare        anaconda/linux-64::pickleshare-0.7.5-py37_1001\n",
      "  prompt-toolkit     anaconda/noarch::prompt-toolkit-3.0.8-py_0\n",
      "  ptyprocess         anaconda/linux-64::ptyprocess-0.6.0-py37_0\n",
      "  pygments           anaconda/noarch::pygments-2.7.1-py_0\n",
      "  python-dateutil    anaconda/noarch::python-dateutil-2.8.1-py_0\n",
      "  pyzmq              anaconda/linux-64::pyzmq-19.0.2-py37he6710b0_1\n",
      "  six                anaconda/noarch::six-1.15.0-py_0\n",
      "  tornado            anaconda/linux-64::tornado-6.0.4-py37h7b6447c_1\n",
      "  traitlets          anaconda/noarch::traitlets-5.0.5-py_0\n",
      "  wcwidth            anaconda/noarch::wcwidth-0.2.5-py_0\n",
      "  zeromq             anaconda/linux-64::zeromq-4.3.3-he6710b0_3\n",
      "\n",
      "The following packages will be SUPERSEDED by a higher-priority channel:\n",
      "\n",
      "  ca-certificates    conda-forge::ca-certificates-2021.10.~ --> anaconda::ca-certificates-2020.10.14-0\n",
      "\n",
      "\n",
      "Preparing transaction: ...working... done\n",
      "Verifying transaction: ...working... done\n",
      "Executing transaction: ...working... done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "  current version: 4.11.0\n",
      "  latest version: 4.12.0\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base conda\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# replace prokka_env with the name of your environment\n",
    "# replace prokka with whatever package you want to download\n",
    "# this may take several minutes, depending on the package\n",
    "ENV_NAME=prokka_env\n",
    "conda create -y -n $ENV_NAME -c conda-forge -c bioconda -c defaults prokka\n",
    "conda install -y -n $ENV_NAME -c anaconda ipykernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0f5c198a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reload your page, go to the Kernel menu, and choose \"Change Kernel\"\n",
    "# your new environment should show up; select it and wait for it to start\n",
    "# you should now have access to the tools you installed \n",
    "\n",
    "# IMPORTANT NOTE: some packages do need additional prerequisites not installed via conda.\n",
    "# If you receive unexpected behavior, first check the Github/Biostars/Conda pages for open issues\n",
    "# You may also need to try installing specific versions of programs/channels, i.e.,\n",
    "#   conda create -n roary_env -c bioconda roary -- DOESN'T WORK (fails when trying to run roary)\n",
    "#   conda create -n roary_env -c conda-forge -c bioconda roary=3.13 -- DOES WORK (but takes a while)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d7ff7dc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Option h is ambiguous (help, hmms)\r\n",
      "Name:\r\n",
      "  Prokka 1.14.6 by Torsten Seemann <torsten.seemann@gmail.com>\r\n",
      "Synopsis:\r\n",
      "  rapid bacterial genome annotation\r\n",
      "Usage:\r\n",
      "  prokka [options] <contigs.fasta>\r\n",
      "General:\r\n",
      "  --help             This help\r\n",
      "  --version          Print version and exit\r\n",
      "  --citation         Print citation for referencing Prokka\r\n",
      "  --quiet            No screen output (default OFF)\r\n",
      "  --debug            Debug mode: keep all temporary files (default OFF)\r\n",
      "Setup:\r\n",
      "  --dbdir [X]        Prokka database root folders (default '/root/miniconda3/envs/prokka_env/db')\r\n",
      "  --listdb           List all configured databases\r\n",
      "  --setupdb          Index all installed databases\r\n",
      "  --cleandb          Remove all database indices\r\n",
      "  --depends          List all software dependencies\r\n",
      "Outputs:\r\n",
      "  --outdir [X]       Output folder [auto] (default '')\r\n",
      "  --force            Force overwriting existing output folder (default OFF)\r\n",
      "  --prefix [X]       Filename output prefix [auto] (default '')\r\n",
      "  --addgenes         Add 'gene' features for each 'CDS' feature (default OFF)\r\n",
      "  --addmrna          Add 'mRNA' features for each 'CDS' feature (default OFF)\r\n",
      "  --locustag [X]     Locus tag prefix [auto] (default '')\r\n",
      "  --increment [N]    Locus tag counter increment (default '1')\r\n",
      "  --gffver [N]       GFF version (default '3')\r\n",
      "  --compliant        Force Genbank/ENA/DDJB compliance: --addgenes --mincontiglen 200 --centre XXX (default OFF)\r\n",
      "  --centre [X]       Sequencing centre ID. (default '')\r\n",
      "  --accver [N]       Version to put in Genbank file (default '1')\r\n",
      "Organism details:\r\n",
      "  --genus [X]        Genus name (default 'Genus')\r\n",
      "  --species [X]      Species name (default 'species')\r\n",
      "  --strain [X]       Strain name (default 'strain')\r\n",
      "  --plasmid [X]      Plasmid name or identifier (default '')\r\n",
      "Annotations:\r\n",
      "  --kingdom [X]      Annotation mode: Archaea|Bacteria|Mitochondria|Viruses (default 'Bacteria')\r\n",
      "  --gcode [N]        Genetic code / Translation table (set if --kingdom is set) (default '0')\r\n",
      "  --prodigaltf [X]   Prodigal training file (default '')\r\n",
      "  --gram [X]         Gram: -/neg +/pos (default '')\r\n",
      "  --usegenus         Use genus-specific BLAST databases (needs --genus) (default OFF)\r\n",
      "  --proteins [X]     FASTA or GBK file to use as 1st priority (default '')\r\n",
      "  --hmms [X]         Trusted HMM to first annotate from (default '')\r\n",
      "  --metagenome       Improve gene predictions for highly fragmented genomes (default OFF)\r\n",
      "  --rawproduct       Do not clean up /product annotation (default OFF)\r\n",
      "  --cdsrnaolap       Allow [tr]RNA to overlap CDS (default OFF)\r\n",
      "Matching:\r\n",
      "  --evalue [n.n]     Similarity e-value cut-off (default '1e-09')\r\n",
      "  --coverage [n.n]   Minimum coverage on query protein (default '80')\r\n",
      "Computation:\r\n",
      "  --cpus [N]         Number of CPUs to use [0=all] (default '8')\r\n",
      "  --fast             Fast mode - only use basic BLASTP databases (default OFF)\r\n",
      "  --noanno           For CDS just set /product=\"unannotated protein\" (default OFF)\r\n",
      "  --mincontiglen [N] Minimum contig size [NCBI needs 200] (default '1')\r\n",
      "  --rfam             Enable searching for ncRNAs with Infernal+Rfam (SLOW!) (default '0')\r\n",
      "  --norrna           Don't run rRNA search (default OFF)\r\n",
      "  --notrna           Don't run tRNA search (default OFF)\r\n",
      "  --rnammer          Prefer RNAmmer over Barrnap for rRNA prediction (default OFF)\r\n"
     ]
    }
   ],
   "source": [
    "!prokka -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2bc54b9b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in PROKKA_04102022/PROKKA_04102022.gbf\n",
      "[tbl2asn-forever] Correcting dates in PROKKA_04102022/PROKKA_04102022.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 10-APR-2022\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[13:49:18] This is prokka 1.14.6\n",
      "[13:49:18] Written by Torsten Seemann <torsten.seemann@gmail.com>\n",
      "[13:49:18] Homepage is https://github.com/tseemann/prokka\n",
      "[13:49:18] Local time is Sun Apr 10 13:49:18 2022\n",
      "[13:49:18] You are not telling me who you are!\n",
      "[13:49:18] Operating system is linux\n",
      "[13:49:18] You have BioPerl 1.7.8\n",
      "Argument \"1.7.8\" isn't numeric in numeric lt (<) at /root/miniconda3/envs/prokka_env/bin/prokka line 259.\n",
      "[13:49:18] System has 12 cores.\n",
      "[13:49:18] Will use maximum of 8 cores.\n",
      "[13:49:18] Annotating as >>> Bacteria <<<\n",
      "[13:49:18] Generating locus_tag from 'lab2/q2.fasta' contents.\n",
      "[13:49:18] Setting --locustag CFFFKGGP from MD5 cfff4009a9292aa64a59e5a4fe4d6f3b\n",
      "[13:49:18] Re-using existing --outdir PROKKA_04102022\n",
      "[13:49:18] Using filename prefix: PROKKA_04102022.XXX\n",
      "[13:49:18] Setting HMMER_NCPU=1\n",
      "[13:49:18] Writing log to: PROKKA_04102022/PROKKA_04102022.log\n",
      "[13:49:18] Command: /root/miniconda3/envs/prokka_env/bin/prokka --force lab2/q2.fasta\n",
      "[13:49:18] Appending to PATH: /root/miniconda3/envs/prokka_env/bin\n",
      "[13:49:18] Looking for 'aragorn' - found /root/miniconda3/envs/prokka_env/bin/aragorn\n",
      "[13:49:18] Determined aragorn version is 001002 from 'ARAGORN v1.2.38 Dean Laslett'\n",
      "[13:49:18] Looking for 'barrnap' - found /root/miniconda3/envs/prokka_env/bin/barrnap\n",
      "[13:49:18] Determined barrnap version is 000009 from 'barrnap 0.9'\n",
      "[13:49:18] Looking for 'blastp' - found /root/miniconda3/envs/prokka_env/bin/blastp\n",
      "[13:49:18] Determined blastp version is 002012 from 'blastp: 2.12.0+'\n",
      "[13:49:18] Looking for 'cmpress' - found /root/miniconda3/envs/prokka_env/bin/cmpress\n",
      "[13:49:18] Determined cmpress version is 001001 from '# INFERNAL 1.1.4 (Dec 2020)'\n",
      "[13:49:18] Looking for 'cmscan' - found /root/miniconda3/envs/prokka_env/bin/cmscan\n",
      "[13:49:18] Determined cmscan version is 001001 from '# INFERNAL 1.1.4 (Dec 2020)'\n",
      "[13:49:18] Looking for 'egrep' - found /usr/bin/egrep\n",
      "[13:49:18] Looking for 'find' - found /usr/bin/find\n",
      "[13:49:18] Looking for 'grep' - found /usr/bin/grep\n",
      "[13:49:18] Looking for 'hmmpress' - found /root/miniconda3/envs/prokka_env/bin/hmmpress\n",
      "[13:49:18] Determined hmmpress version is 003003 from '# HMMER 3.3.2 (Nov 2020); http://hmmer.org/'\n",
      "[13:49:18] Looking for 'hmmscan' - found /root/miniconda3/envs/prokka_env/bin/hmmscan\n",
      "[13:49:18] Determined hmmscan version is 003003 from '# HMMER 3.3.2 (Nov 2020); http://hmmer.org/'\n",
      "[13:49:18] Looking for 'java' - found /root/miniconda3/envs/prokka_env/bin/java\n",
      "[13:49:18] Looking for 'makeblastdb' - found /root/miniconda3/envs/prokka_env/bin/makeblastdb\n",
      "[13:49:18] Determined makeblastdb version is 002012 from 'makeblastdb: 2.12.0+'\n",
      "[13:49:18] Looking for 'minced' - found /root/miniconda3/envs/prokka_env/bin/minced\n",
      "[13:49:18] Determined minced version is 004002 from 'minced 0.4.2'\n",
      "[13:49:18] Looking for 'parallel' - found /root/miniconda3/envs/prokka_env/bin/parallel\n",
      "[13:49:18] Determined parallel version is 20220222 from 'GNU parallel 20220222'\n",
      "[13:49:18] Looking for 'prodigal' - found /root/miniconda3/envs/prokka_env/bin/prodigal\n",
      "[13:49:18] Determined prodigal version is 002006 from 'Prodigal V2.6.3: February, 2016'\n",
      "[13:49:18] Looking for 'prokka-genbank_to_fasta_db' - found /root/miniconda3/envs/prokka_env/bin/prokka-genbank_to_fasta_db\n",
      "[13:49:18] Looking for 'sed' - found /usr/bin/sed\n",
      "[13:49:18] Looking for 'tbl2asn' - found /root/miniconda3/envs/prokka_env/bin/tbl2asn\n",
      "[13:49:18] Determined tbl2asn version is 025007 from 'tbl2asn 25.7   arguments:'\n",
      "[13:49:18] Using genetic code table 11.\n",
      "[13:49:18] Loading and checking input file: lab2/q2.fasta\n",
      "[13:49:18] Wrote 3 contigs totalling 3582 bp.\n",
      "[13:49:18] Predicting tRNAs and tmRNAs\n",
      "[13:49:18] Running: aragorn -l -gc11  -w PROKKA_04102022\\/PROKKA_04102022\\.fna\n",
      "[13:49:18] Found 0 tRNAs\n",
      "[13:49:18] Predicting Ribosomal RNAs\n",
      "[13:49:18] Running Barrnap with 8 threads\n",
      "[13:49:18] Found 0 rRNAs\n",
      "[13:49:18] Skipping ncRNA search, enable with --rfam if desired.\n",
      "[13:49:18] Total of 0 tRNA + rRNA features\n",
      "[13:49:18] Searching for CRISPR repeats\n",
      "[13:49:18] Found 0 CRISPRs\n",
      "[13:49:18] Predicting coding sequences\n",
      "[13:49:18] Contigs total 3582 bp, so using meta mode\n",
      "[13:49:18] Running: prodigal -i PROKKA_04102022\\/PROKKA_04102022\\.fna -c -m -g 11 -p meta -f sco -q\n",
      "[13:49:18] Found 3 CDS\n",
      "[13:49:18] Connecting features back to sequences\n",
      "[13:49:18] Not using genus-specific database. Try --usegenus to enable it.\n",
      "[13:49:18] Annotating CDS, please be patient.\n",
      "[13:49:18] Will use 8 CPUs for similarity searching.\n",
      "[13:49:18] There are still 3 unannotated CDS left (started with 3)\n",
      "[13:49:18] Will use blast to search against /root/miniconda3/envs/prokka_env/db/kingdom/Bacteria/IS with 8 CPUs\n",
      "[13:49:18] Running: cat PROKKA_04102022\\/PROKKA_04102022\\.IS\\.tmp\\.20140\\.faa | parallel --gnu --plain -j 8 --block 68 --recstart '>' --pipe blastp -query - -db /root/miniconda3/envs/prokka_env/db/kingdom/Bacteria/IS -evalue 1e-30 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > PROKKA_04102022\\/PROKKA_04102022\\.IS\\.tmp\\.20140\\.blast 2> /dev/null\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.IS.tmp.20140.faa\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.IS.tmp.20140.blast\n",
      "[13:49:19] There are still 3 unannotated CDS left (started with 3)\n",
      "[13:49:19] Will use blast to search against /root/miniconda3/envs/prokka_env/db/kingdom/Bacteria/AMR with 8 CPUs\n",
      "[13:49:19] Running: cat PROKKA_04102022\\/PROKKA_04102022\\.AMR\\.tmp\\.20140\\.faa | parallel --gnu --plain -j 8 --block 68 --recstart '>' --pipe blastp -query - -db /root/miniconda3/envs/prokka_env/db/kingdom/Bacteria/AMR -evalue 1e-300 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > PROKKA_04102022\\/PROKKA_04102022\\.AMR\\.tmp\\.20140\\.blast 2> /dev/null\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.AMR.tmp.20140.faa\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.AMR.tmp.20140.blast\n",
      "[13:49:19] There are still 3 unannotated CDS left (started with 3)\n",
      "[13:49:19] Will use blast to search against /root/miniconda3/envs/prokka_env/db/kingdom/Bacteria/sprot with 8 CPUs\n",
      "[13:49:19] Running: cat PROKKA_04102022\\/PROKKA_04102022\\.sprot\\.tmp\\.20140\\.faa | parallel --gnu --plain -j 8 --block 68 --recstart '>' --pipe blastp -query - -db /root/miniconda3/envs/prokka_env/db/kingdom/Bacteria/sprot -evalue 1e-09 -qcov_hsp_perc 80 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > PROKKA_04102022\\/PROKKA_04102022\\.sprot\\.tmp\\.20140\\.blast 2> /dev/null\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.sprot.tmp.20140.faa\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.sprot.tmp.20140.blast\n",
      "[13:49:19] Found 1 unique /gene codes.\n",
      "[13:49:19] Fixed 3 duplicate /gene - acrA_1 acrA_2 acrA_3\n",
      "[13:49:19] Fixed 1 colliding /gene names.\n",
      "[13:49:19] Adding /locus_tag identifiers\n",
      "[13:49:19] Assigned 3 locus_tags to CDS and RNA features.\n",
      "[13:49:19] Writing outputs to PROKKA_04102022/\n",
      "[13:49:19] Generating annotation statistics file\n",
      "[13:49:19] Generating Genbank and Sequin files\n",
      "[13:49:19] Running: tbl2asn -V b -a r10k -l paired-ends -M n -N 1 -y 'Annotated using prokka 1.14.6 from https://github.com/tseemann/prokka' -Z PROKKA_04102022\\/PROKKA_04102022\\.err -i PROKKA_04102022\\/PROKKA_04102022\\.fsa 2> /dev/null\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/errorsummary.val\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.dr\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.fixedproducts\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.ecn\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.val\n",
      "[13:49:19] Repairing broken .GBK output that tbl2asn produces...\n",
      "[13:49:19] Running: sed 's/COORDINATES: profile/COORDINATES:profile/' < PROKKA_04102022\\/PROKKA_04102022\\.gbf > PROKKA_04102022\\/PROKKA_04102022\\.gbk\n",
      "[13:49:19] Deleting unwanted file: PROKKA_04102022/PROKKA_04102022.gbf\n",
      "[13:49:19] Output files:\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.txt\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.gff\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.fna\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.tbl\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.gbk\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.HAMAP.hmm.tmp.20140.faa\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.log\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.ffn\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.tsv\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.fsa\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.err\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.sqn\n",
      "[13:49:19] PROKKA_04102022/PROKKA_04102022.faa\n",
      "[13:49:19] Annotation finished successfully.\n",
      "[13:49:19] Walltime used: 0.02 minutes\n",
      "[13:49:19] If you use this result please cite the Prokka paper:\n",
      "[13:49:19] Seemann T (2014) Prokka: rapid prokaryotic genome annotation. Bioinformatics. 30(14):2068-9.\n",
      "[13:49:19] Type 'prokka --citation' for more details.\n",
      "[13:49:19] Share and enjoy!\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "prokka --force lab2/q2.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03cf832c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:prokka_env]",
   "language": "python",
   "name": "conda-env-prokka_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
